## Allohexaploid *T. aestivum*

```{r tt-wheat-setup, include=FALSE}
source("../_config.R")
```

### Introduction

Bread wheat, *Triticum aestivum*, is one of the worldâ€™s most important crops
and also serves as a model for studying allopolyploid species.
It originated through hybridization among three diploid progenitor species,
Triticum urartu (AA), a species related to *Aegilops speltoides* (BB),
and *Aegilops tauschii* (DD), and contains three distinct subgenomes
[@ref_wheat_origin_1; @ref_wheat_origin_2; @ref_wheatgenome].
To demonstrate the applicability of HOBIT to allohexaploid species,
we present a case study using wheat RNA-Seq data
to detect homeologs with shifts in expression ratios between two tissues,
shoot apex and leaf [@ref_wheatseedmatrix].


### Data Preparation

In this example, 1,000 homeolog triads were randomly selected from the original dataset.
The expression data can be loaded from the following file (`t_aestivum.w20.mini.txt.gz`).
Users may also choose to load the full dataset, which is stored
in the [data](https://github.com/bitdessin/hespresso/tree/main/docs/data) directory, for analysis.

```{r tt-wheat-load-counts}
gexp <- read.table("../data/t_aestivum.w20.mini.txt.gz", header = TRUE, sep = "\t", row.names = 1)
head(gexp)
group <- c("apex", "apex", "apex", "leaf", "leaf", "leaf")
```

Next, load the mapping table that links each gene to its corresponding homeolog triads.
Since wheat has three subgenomes, the table contains three columns,
with each column representing one subgenome.

```{r tt-wheat-load-mappingtable}
mapping_table <- read.table("../data/t_aestivum.homeolog.txt.gz", header = TRUE, sep = "\t")
head(mapping_table)
```

We then construct an `ExpMX` class object using the `newExpMX()` function.
This converts the gene expression matrix (`gexp`)
into a homeolog expression matrix based on the mapping table (`mapping_table`).

```{r tt-wheat-create-expmx}
x <- newExpMX(gexp, group, mapping_table)
x
```

Before performing the test,
we normalize the raw read counts using the TMM method [@ref_tmm]
to adjust for differences in library size.
However, if the expression data (`gexp`) has already been normalized (e.g. FPKM),
this step can be skipped.

```{r tt-wheat-norm-counts}
x <- norm_counts(x)
```

To visualize the distributions of homeolog expression ratios, use the `plot_HER_distr()` function.
By default, the expression ratio is calculated as the proportion contributed
by the first subgenome relative to all subgenomes.
In this dataset, the first subgenome corresponds to the A-subgenome,
as indicated by the first column of the mapping table (`mapping_table`).


```{r tt-wheat-viz-distr, fig.cap="Distribution of homeolog expression ratios between shoot apex and leaf in the A-subgenome of wheat."}
distr_plots_A <- plot_HER_distr(x)
names(distr_plots_A)

library(ggplot2)
library(gridExtra)

grid.arrange(distr_plots_A[["apex"]] + ggtitle("apex on A subgenome"),
             distr_plots_A[["leaf"]] + ggtitle("leaf on A subgenome"),
             ncol = 2)
```

To calculate expression ratios from the second or third subgenome,
set the `base` option accordingly.
For example, to calculate the ratios for the second and third subgenomes
(i.e., B-subgenome and D-subgenome), set `base = 2` and `base = 3`, respectively.


```{r tt-wheat-viz-distrs, fig.cap="Distribution of homeolog expression ratios between shoot apex and leaf across all subgenomes."}
distr_plots_B <- plot_HER_distr(x, base = 2)
distr_plots_D <- plot_HER_distr(x, base = 3)

grid.arrange(distr_plots_A[["apex"]] + ggtitle("apex on A"), distr_plots_A[["leaf"]] + ggtitle("leaf on A"),
             distr_plots_B[["apex"]] + ggtitle("apex on B"), distr_plots_B[["leaf"]] + ggtitle("leaf on B"),
             distr_plots_D[["apex"]] + ggtitle("apex on D"), distr_plots_D[["leaf"]] + ggtitle("leaf on D"),
             ncol = 2)
```

Histogram peaks at one-third for all subgenomes indicate
that most homeologs in wheat are equally expressed from the three subgenomes.

### Statistical Test

After visually inspecting the distributions of homeolog expression ratios,
use the `hobit()` function to identify homeologs
showing significant changes in expression ratios between shoot apex and leaf tissues.
This step takes approximately five minutes using eight threads and may vary depending on hardware performance.

```{r tt-wheat-run-hobit, results="hide", message=FALSE, warning=FALSE}
x_output <- hobit(x)
```

```{r tt-wheat-hobit-output}
head(x_output)
```

The output can be sorted in ascending order of *p*-values using the following code.

```{r tt-wheat-hobit-sort-output}
head(x_output[order(x_output$pvalue), ])
```

Next, we visualize changes in expression ratios of the first subgenome between the two tissues.
In the plot, homeologs with *q*-values less than 0.01 are highlighted.

```{r tt-wheat-viz-HER-changes, fig.cap="Changes in homeolog expression ratios of the A-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes."}
is_sig <- ifelse(x_output$qvalue < 0.01, "q<0.01", "q>=0.01")
plot_HER(x, label = is_sig)
```

To change the subgenome used for calculating expression ratios,
set the `base` argument to the corresponding subgenome number.
For example, to use the second subgenome, run the following.

```{r tt-wheat-viz-HER-changes-base-2, fig.cap="Changes in homeolog expression ratios of the B-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes."}
plot_HER(x, base = 2, label = is_sig)
```


Note that expression ratios are calculated for the selected subgenome,
while highlighted homeologs are those that meet the significance threshold in any subgenome.
As a result, some homeologs may appear highlighted even if they show only minor changes in the plot,
because they may have significantly altered expression ratios in another subgenome.

### Combining Subgenomes

Additionally, users who want to test expression ratios
between the combined A- and B-subgenomes versus the D-subgenome
can first merge the expression values of the A and B homeologs
using the `combine_hexp()` function.
After combining, the statistical test can be performed on the merged dataset.
Assuming the A- and B-subgenomes are stored as the first and second subgenomes
in the `x` object (as defined in the `mapping_table`),
set `subgenomes = c(1, 2)` to combine them.


```{r tt-wheat-combine-subgenomes}
x_ABvsD <- combine_hexp(x, subgenomes = c(1, 2))
x_ABvsD
```

HOBIT can then be applied to the combined dataset as below.

```{r tt-wheat-combine-subgenomes-hobit, results="hide", message=FALSE, warning=FALSE}
x_ABvsD_output <- hobit(x_ABvsD)
```

```{r tt-wheat-combine-subgenomes-hobit-output}
head(x_ABvsD_output[order(x_ABvsD_output$pvalue), ])
```


### Analysis Environment

```{r tt-wheat-sessioninfo}
sessionInfo()
```
