---
title: "_hespresso_: 3. Simulation Evaluation"
date: "Updated: 2025-07-06 Compiled: `r format(Sys.time(), '%Y-%m-%d')`."
author:
    - name: Jianqiang Sun
      affiliation: |
        Research Center for Agricultural Information Technology,
        National Agriculture and Food Research Organization, Japan
output: 
    BiocStyle::html_document:
        highlight: pygments
        toc: true
        toc_depth: 1
editor_options:
    chunk_output_type: console
bibliography: references.bib
package: hespresso
vignette: |
    %\VignetteIndexEntry{hespresso: 3. Simulation Evaluation}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
abstract: |
    Simulation studies offer a powerful approach for evaluating method performance
    under controlled conditions. The *hespresso* package includes a suite of functions
    to generate artificial datasets that simulate read counts from allopolyploid species.
    These functions enable comprehensive assessment of HOBIT and related methods across
    a range of experimental scenarios.
---

```{r SIM-startup, echo=FALSE, results="hide", message=FALSE}
library(hespresso)
options(mc.cores = 4)
options(ggplot2.discrete.colour = c("#999999", "#E69F00", "#56B4E9", "#009E73",
                                    "#F0E442", "#0072B2", "#D55E00", "#CC79A7"))
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      error = TRUE,
                      warning = FALSE,
                      message = FALSE,
                      dev = "png",
                      fig.width = 4.2,
                      fig.height = 3.0,
                      fig.small = TRUE)
set.seed(1)
```


# Simulating Allotetraploid Data

The `r Biocpkg("hespresso")` package provides functions
to simulate artificial read counts for allotetraploid and allohexaploid species.
By default, the simulation generates read counts for
an allotetraploid with 10,000 homeologs across two conditions,
each with three biological replicates.
The mean and dispersion parameters are estimated from real RNA-Seq data from
an allotetraploid *Cardamine flexuosa* dataset [@ref_cflex_ecohabitats].

To generate a simulated dataset with default options, run as follows.

```{r SIM-2x-default}
x <- sim_homeolog_counts()
x
```

Options can be flexibly customized by users.
For example, to simulate data with 1,000 homeologs and 5 replicates per condition,
run as follows.

```{r SIM-2x-1000g-3r}
x <- sim_homeolog_counts(n_gene = 1000, n_replicates = c(5, 5))
x
```

Simulated read counts are generated from a **negative binomial** (**NB**) distribution.
The mean and dispersion values used for each homeolog are stored in the object returned by `sim_homeolog_counts()`.
The `get_sim_params()` function can be used to retrieve these parameters.
For example, to get means used for sampling read counts, run as follows.

```{r SIM-2x-getsimparams}
param_mu <- get_sim_params(x, "mu")
```

The output is a list of DataFrame objects, with each element corresponding to a condition group.
Each DataFrame contains the true mean expression values used for sampling,
where the columns represent individual subgenomes.
These ground-truth mean values can be used to calculate the true
**homeolog expression ratios** (**HERs**) for each condition.
Based on these HERs, users can set thresholds to define homeologs that shift
HERs significantly, referred to as **ratio-shifted homeologs** (**RSHs**), between conditions.
 

```{r SIM-2x-getsimparams-output}
names(param_mu)
head(param_mu[["group_1"]])
```

Alternatively, users can use the `def_sigShift()` function to automatically
define RSHs without manually calculating summary statistics.
For example, to identify RSHs as those with a maximum absolute
HER difference (`Dmax`) greater than 0.2 or a maximum odds ratio (`ORmax`)
exceeding 2.0, use the following command.

```{r SIM-2x-def-gtRCH}
x <- sim_homeolog_counts()

is_sig <- def_sigShift(x, Dmax = 0.2, ORmax = 2, operator = "OR")
table(is_sig)
```

The output indicates that there are `r sum(is_sig)` RSHs defined with the specific thresholds.

To visualize the distribution of HERs across conditions,
use the `plot_HER_distr()` function.
This function returns a list of histograms,
one for each group, showing HER distributions.
By default, HERs of the first subgenome (`base = 1`) are computed.

```{r SIM-2x-viz-run-plotHERdistr}
x <- sim_homeolog_counts()
distr_plots <- plot_HER_distr(x)
```

Since the simulated dataset contains two groups,
the `plot_HER_distr()` function returns two histograms.

```{r SIM-2x-viz-plotnames}
names(distr_plots)
```

To show the HER distribution for each group, run as follows.

```{r SIM-2x-viz-distrs, fig.cap="Distribution of Homeolog Expression Ratios. Histograms display the distributions of homeolog expression ratios under two simulated groups.", fig.height=2.6}
library(ggplot2)
library(gridExtra)

grid.arrange(distr_plots[["group_1"]] + ggtitle("group_1"), 
             distr_plots[["group_2"]] + ggtitle("group_2"),
             ncol = 2)
```


To compare HER changes between groups, user can use the `plot_HER()` function.
By default, it plots HERs for the first subgenome between two groups using a scatter plot.

```{r SIM-2x-viz-plotHER, fig.cap="Changes in Homeolog Expression Ratios Between Two Groups. Each point represents the homeolog expression ratio in group 1 (x-axis) and group 2 (y-axis)."}
plot_HER(x, alpha = 0.3)
```

If the dataset was generated using `sim_homeolog_counts()`,
user can incorporate the `def_sigShift()` function
to highlight ground-true RSHs in the plot.

```{r SIM-2x-viz-plotHER-labels, fig.cap="Changes in Homeolog Expression Ratios Between Two Groups. Orange points indicate homeologs with significant shifts in expression ratios between the two groups, whereas gray points represent homeologs without significant changes."}
is_sig <- def_sigShift(x, Dmax = 0.2, ORmax = 2.0)
plot_HER(x, label = is_sig, alpha = 0.3)
```


# Simulating Allohexaploid Data

To simulate read counts for an allohexaploid with three subgenomes, set `n_subgenomes = 3`.
In this case, the mean and dispersion parameters are derived
from real RNA-Seq data from allohexaploid wheat [@ref_wheatseedmatrix].

```{r SIM-3x}
x <- sim_homeolog_counts(n_subgenomes = 3)
```

The simulation parameters can be retrieved using the `get_sim_params()` function
and the ground-truth RSHs can be defined with the `def_sigShift()` function,
as like example showing for allotetraploid data.

For exmaple, using `def_sigShift()` function can define ground-truth RSHs,
homeologs with large changes satisfied the given thresholds in any subgenome.

```{r SIM-3x-def-gtRCH-hex}
is_sig <- def_sigShift(x, Dmax = 0.2, ORmax = 2, operator = "OR")
table(is_sig)
```

Alternatively, to define RSHs specific to a single subgenome,
for example, only the first subgenome, set the `base` argument accordingly as follows.

```{r SIM-simdata-def-gtRCH-hex-base-1}
x <- sim_homeolog_counts(n_subgenomes = 3)

is_sig <- def_sigShift(x, base = 1, Dmax = 0.2, ORmax = 2, operator = "OR")
table(is_sig)
```

Visualization for allohexaploid data, comprising three subgenomes, works similarly.

```{r SIM-viz-distr-simdata-6x, fig.cap="Distribution of Homeolog Expression Ratios for the First Subgenome. Histograms illustrate the distributions of homeolog expression ratios for the first subgenome in a simulated allohexaploid dataset across two experimental groups.", fig.height=2.6}
x <- sim_homeolog_counts(n_subgenomes = 3)
distr_plots_1 <- plot_HER_distr(x)
names(distr_plots_1)

library(ggplot2)
library(gridExtra)
grid.arrange(distr_plots_1[["group_1"]] + ggtitle("group_1 on 1st subgenome"), 
             distr_plots_1[["group_2"]] + ggtitle("group_2 on 1st subgenome"),
             ncol = 2)
```

User can specify which subgenome to use for HER calculations with the `base` argument.
For example, to visualize HERs for the second subgenome.

```{r SIM-viz-3x-her-distrs, fig.cap="Distribution of Homeolog Expression Ratios Across All Subgenomes. Histograms show the distributions of homeolog expression ratios for the first, second, and third subgenomes in a simulated allohexaploid dataset across two groups.", fig.height=4.8}
distr_plots_2 <- plot_HER_distr(x, base = 2)
distr_plots_3 <- plot_HER_distr(x, base = 3)

grid.arrange(distr_plots_1[["group_1"]] + ggtitle("group_1 on #1"), distr_plots_1[["group_2"]] + ggtitle("group_2 on #1"),
             distr_plots_2[["group_1"]] + ggtitle("group_1 on #2"), distr_plots_2[["group_2"]] + ggtitle("group_2 on #2"),
             distr_plots_3[["group_1"]] + ggtitle("group_1 on #3"), distr_plots_3[["group_2"]] + ggtitle("group_2 on #3"),
             ncol = 2)
```

The `plot_HER()` function also works for allohexaploids.
By default, HERs are computed from the first subgenome.

```{r SIM-viz-her-3x-changing, fig.cap="Changes in Homeolog Expression Ratios of the First Subgenome Between Two Groups. Each point represents the homeolog expression ratio of the first subgenome in two groups."}
plot_HER(x, alpha = 0.3)
```

To specify which subgenome to use for HER calculations, user can use the `base` argument.
The following example visualizes HERs of the second subgenome between two groups.

```{r SIM-viz-her-3x-changing-base-2, fig.cap="Changes in Homeolog Expression Ratios of the Second Subgenome Between Two Groups. Each point represents the homeolog expression ratio of the second subgenome in two groups."}
plot_HER(x, base = 2, alpha = 0.3)
```

In addition, with simulated data, user can incorporate the `def_sigShift()` function
to obtain the ground-truth RSHs and highlight them in the scatter plot.
The following example visualizes HERs of the first subgenome between two groups
and highlights homeologs that satisfy the RSH thresholds in any subgenome.
Since the RSHs are defined across all subgenomes
while the plot only shows the first subgenome,
homeologs with minimal changes in the first subgenome may still appear highlighted.

```{r SIM-viz-3x-herchanging-labels, fig.cap="Changes in Homeolog Expression Ratios of the First Subgenome Between Two Groups. Orange points represent homeologs with significant changes in expression ratios between the two groups for any subgenome, while gray points indicate homeologs without significant changes."}
is_sig <- def_sigShift(x, base = 0, Dmax = 0.2, ORmax = 2)
plot_HER(x, label = is_sig, alpha = 0.3)
```

Alternatively, to highlight only those homeologs
that meet the RSH threshold specifically in the first subgenome, use as the follows.

```{r SIM-viz-3x-herchanging-labels-base-2, fig.cap="Changes in Homeolog Expression Ratios of the First Subgenome Between Two Groups. Orange points represent homeologs with significant changes in expression ratios of the first subgenome between the two groups, while gray points indicate homeologs without significant changes."}
is_sig_b1 <- def_sigShift(x, base = 1, Dmax = 0.2, ORmax = 1.8)
plot_HER(x, base = 1, label = is_sig_b1, alpha = 0.3)
```



# Simulating from a Custom Expression Seed

Instead of using the default datasets based on *C. flexuosa* or wheat,
users can supply their own expression matrix
to calculate the mean and dispersion parameters for simulation.
The matrix should be formatted with genes as rows and samples (biological replicates) as columns.
Each cell should contain a normalized read count.

Below is an example showing
how to load a custom matrix and use it as a template for simulation.

```{r SIM-seedmatrix}
expmx <- read.table(system.file(package = "hespresso", "extdata", "seed_matrix.C_flexuosa.tsv.gz"))
head(expmx)
x <- sim_homeolog_counts(seed_expmx = expmx)
x
```

This function allows users to evaluate HOBIT performance
on data representative of their own study system.


# Performance Evaluation

By combining the `sim_homeolog_counts()` and `def_sigShift()` functions,
users can evaluate the performance of HOBIT using standard metrics
such as precision, recall, F1 score, and area under the ROC curve (AUC).

For a reliable evaluation of HOBIT,
we recommend simulating a large number of homeologs to ensure the dataset
includes a sufficient number of RSHs for robust validation.
However, to reduce computation time in this example,
we simulate data for only 100 homeologs for performance evaluation.


```{r SIM-eval-simdata, results="hide", message=FALSE, warning=FALSE}
x <- sim_homeolog_counts(n_genes = 100)
x_output <- hobit(x)
```

Next, we identify predicted RSHs using an FDR threshold of 0.05.
We then calculate precision, recall, and the F1 score
by comparing the predicted RSHs against the ground-truth RSHs,
which are defined using the `def_sigShift()` function.
To compute the AUC, we use the *p*-values from the `hobit()` output
and input them into the `auc()` and `roc()` functions provided by the `r Biocpkg("pROC")` package.

```{r SIM-eval-calcmetrics}
library(pROC)

pred_sig <- (x_output$qvalue < 0.05)
true_sig <- def_sigShift(x, Dmax = 0.2, ORmax = 2, operator = "OR")

tp <- sum(pred_sig & true_sig)
tn <- sum(!pred_sig & !true_sig)
fp <- sum(pred_sig & !true_sig)
fn <- sum(!pred_sig & true_sig)

precision <- tp / (tp + fp)
recall <- tp / (tp + fn)
f1 <- 2 * precision * recall / (precision + recall)
auc <- auc(roc(true_sig, 1 - x_output$pvalue, levels = c(FALSE, TRUE), direction = "<"))

print(c(precision = precision, recall = recall, f1 = f1, auc = auc))
```




# Session Information

```{r SIM-sessionInfo}
sessionInfo()
```



# References


