<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.2 Allohexaploid T. aestivum | hespresso ☕</title>
  <meta name="description" content="3.2 Allohexaploid T. aestivum | hespresso ☕" />
  <meta name="generator" content="bookdown 0.44 and GitBook 2.6.7" />

  <meta property="og:title" content="3.2 Allohexaploid T. aestivum | hespresso ☕" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.2 Allohexaploid T. aestivum | hespresso ☕" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="allohexaploid-c.-flexuosa.html"/>
<link rel="next" href="allotriploid-c.-insueta.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="_style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>hespresso</a></li>
<li class="chapter" data-level="1" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>1</b> Installation</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a></li>
<li class="chapter" data-level="3" data-path="tutorials.html"><a href="tutorials.html"><i class="fa fa-check"></i><b>3</b> Tutorials</a>
<ul>
<li class="chapter" data-level="3.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html"><i class="fa fa-check"></i><b>3.1</b> Allohexaploid <em>C. flexuosa</em></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#introduction"><i class="fa fa-check"></i><b>3.1.1</b> Introduction</a></li>
<li class="chapter" data-level="3.1.2" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#data-preparation"><i class="fa fa-check"></i><b>3.1.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.1.3" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#hobit"><i class="fa fa-check"></i><b>3.1.3</b> HOBIT</a></li>
<li class="chapter" data-level="3.1.4" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#homeoroq"><i class="fa fa-check"></i><b>3.1.4</b> HomeoRoq</a></li>
<li class="chapter" data-level="3.1.5" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#analysis-environment"><i class="fa fa-check"></i><b>3.1.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html"><i class="fa fa-check"></i><b>3.2</b> Allohexaploid <em>T. aestivum</em></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#introduction-1"><i class="fa fa-check"></i><b>3.2.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#data-preparation-1"><i class="fa fa-check"></i><b>3.2.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.2.3" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#statistical-test"><i class="fa fa-check"></i><b>3.2.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.2.4" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#combining-subgenomes"><i class="fa fa-check"></i><b>3.2.4</b> Combining Subgenomes</a></li>
<li class="chapter" data-level="3.2.5" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#analysis-environment-1"><i class="fa fa-check"></i><b>3.2.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html"><i class="fa fa-check"></i><b>3.3</b> Allotriploid <em>C. insueta</em></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#introduction-2"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#data-preparation-2"><i class="fa fa-check"></i><b>3.3.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#statistical-test-1"><i class="fa fa-check"></i><b>3.3.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.3.4" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#analysis-environment-2"><i class="fa fa-check"></i><b>3.3.4</b> Analysis Environment</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4</b> Functions</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hobit-1.html"><a href="hobit-1.html"><i class="fa fa-check"></i><b>4.1</b> HOBIT</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="hobit-1.html"><a href="hobit-1.html#inputs"><i class="fa fa-check"></i><b>4.1.1</b> Inputs</a></li>
<li class="chapter" data-level="4.1.2" data-path="hobit-1.html"><a href="hobit-1.html#test"><i class="fa fa-check"></i><b>4.1.2</b> Test</a></li>
<li class="chapter" data-level="4.1.3" data-path="hobit-1.html"><a href="hobit-1.html#outputs"><i class="fa fa-check"></i><b>4.1.3</b> Outputs</a></li>
<li class="chapter" data-level="4.1.4" data-path="hobit-1.html"><a href="hobit-1.html#options"><i class="fa fa-check"></i><b>4.1.4</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html"><i class="fa fa-check"></i><b>4.2</b> HomeoRoq</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="homeoroq-1.html"><a href="homeoroq-1.html#inputs-1"><i class="fa fa-check"></i><b>4.2.1</b> Inputs</a></li>
<li class="chapter" data-level="4.2.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html#test-1"><i class="fa fa-check"></i><b>4.2.2</b> Test</a></li>
<li class="chapter" data-level="4.2.3" data-path="homeoroq-1.html"><a href="homeoroq-1.html#outputs-1"><i class="fa fa-check"></i><b>4.2.3</b> Outputs</a></li>
<li class="chapter" data-level="4.2.4" data-path="homeoroq-1.html"><a href="homeoroq-1.html#options-1"><i class="fa fa-check"></i><b>4.2.4</b> Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="evaluation-framework.html"><a href="evaluation-framework.html"><i class="fa fa-check"></i><b>5</b> Evaluation Framework</a>
<ul>
<li class="chapter" data-level="5.1" data-path="simulating-allotetraploid-data.html"><a href="simulating-allotetraploid-data.html"><i class="fa fa-check"></i><b>5.1</b> Simulating Allotetraploid Data</a></li>
<li class="chapter" data-level="5.2" data-path="simulating-allohexaploid-data.html"><a href="simulating-allohexaploid-data.html"><i class="fa fa-check"></i><b>5.2</b> Simulating Allohexaploid Data</a></li>
<li class="chapter" data-level="5.3" data-path="custom-data-simulation.html"><a href="custom-data-simulation.html"><i class="fa fa-check"></i><b>5.3</b> Custom Data Simulation</a></li>
<li class="chapter" data-level="5.4" data-path="evaluation-framework-1.html"><a href="evaluation-framework-1.html"><i class="fa fa-check"></i><b>5.4</b> Evaluation Framework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="citations.html"><a href="citations.html"><i class="fa fa-check"></i>Citations</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><em>hespresso</em> ☕</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="allohexaploid-t.-aestivum" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Allohexaploid <em>T. aestivum</em><a href="allohexaploid-t.-aestivum.html#allohexaploid-t.-aestivum" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="introduction-1" class="section level3 hasAnchor" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Introduction<a href="allohexaploid-t.-aestivum.html#introduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Bread wheat, <em>Triticum aestivum</em>, is one of the world’s most important crops
and also serves as a model for studying allopolyploid species.
It originated through hybridization among three diploid progenitor species,
Triticum urartu (AA), a species related to <em>Aegilops speltoides</em> (BB),
and <em>Aegilops tauschii</em> (DD), and contains three distinct subgenomes
<span class="citation">(IWGSC 2018; Levy and Feldman 2022; Zhao et al. 2023)</span>.
To demonstrate the applicability of HOBIT to allohexaploid species,
we present a case study using wheat RNA-Seq data
to detect homeologs with shifts in expression ratios between two tissues,
shoot apex and leaf <span class="citation">(Yang et al. 2021)</span>.</p>
</div>
<div id="data-preparation-1" class="section level3 hasAnchor" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Data Preparation<a href="allohexaploid-t.-aestivum.html#data-preparation-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this example, 1,000 homeolog triads were randomly selected from the original dataset.
The expression data can be loaded from the following file (<code>t_aestivum.w20.mini.txt.gz</code>).
Users may also choose to load the full dataset, which is stored
in the <a href="https://github.com/bitdessin/hespresso/tree/main/docs/data">data</a> directory, for analysis.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="allohexaploid-t.-aestivum.html#cb30-1" tabindex="-1"></a>gexp <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/t_aestivum.w20.mini.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb30-2"><a href="allohexaploid-t.-aestivum.html#cb30-2" tabindex="-1"></a><span class="fu">head</span>(gexp)</span></code></pre></div>
<pre><code>##                    shootapex_1 shootapex_2 shootapex_3 leaf_1 leaf_2 leaf_3
## TraesCS1A02G009000         256         245         255    118    215    200
## TraesCS1A02G011200          22          21          31     24     24     12
## TraesCS1A02G033800          13          10           9      0      1      0
## TraesCS1A02G038400         990         917         975      5      3      7
## TraesCS1A02G042500         203         207         183      0      0      1
## TraesCS1A02G045800          10          11          20    222    408    313</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="allohexaploid-t.-aestivum.html#cb32-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;apex&quot;</span>, <span class="st">&quot;apex&quot;</span>, <span class="st">&quot;apex&quot;</span>, <span class="st">&quot;leaf&quot;</span>, <span class="st">&quot;leaf&quot;</span>, <span class="st">&quot;leaf&quot;</span>)</span></code></pre></div>
<p>Next, load the mapping table that links each gene to its corresponding homeolog triads.
Since wheat has three subgenomes, the table contains three columns,
with each column representing one subgenome.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="allohexaploid-t.-aestivum.html#cb33-1" tabindex="-1"></a>mapping_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/t_aestivum.homeolog.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb33-2"><a href="allohexaploid-t.-aestivum.html#cb33-2" tabindex="-1"></a><span class="fu">head</span>(mapping_table)</span></code></pre></div>
<pre><code>##                    A                    B                    D
## 1 TraesCS1A02G000800 TraesCS6B02G173200LC   TraesCS5D02G023700
## 2 TraesCS1A02G000900   TraesCS6B02G117800   TraesCS7D02G119300
## 3 TraesCS1A02G001600   TraesCS1B02G320300 TraesCS7D02G148100LC
## 4 TraesCS1A02G001900   TraesCS1B02G000800   TraesCS1D02G003900
## 5 TraesCS1A02G002000   TraesCS1B02G000900   TraesCS1D02G004000
## 6 TraesCS1A02G002100   TraesCS1B02G001000   TraesCS1D02G004100</code></pre>
<p>We then construct an <code>ExpMX</code> class object using the <code>newExpMX()</code> function.
This converts the gene expression matrix (<code>gexp</code>)
into a homeolog expression matrix based on the mapping table (<code>mapping_table</code>).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="allohexaploid-t.-aestivum.html#cb35-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">newExpMX</span>(gexp, group, mapping_table)</span>
<span id="cb35-2"><a href="allohexaploid-t.-aestivum.html#cb35-2" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>## # 3 subgenomes (A, B, D)
## # 1000 homeologs
## ---------------------
## Experiment Design:
##   group
## 1  apex
## 2  apex
## 3  apex
## 4  leaf
## 5  leaf
## 6  leaf
## ---------------------
## &gt; subgenome: A 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]  256  245  255  118  215  200
## [2,]   22   21   31   24   24   12
## [3,]   13   10    9    0    1    0
## [4,]  990  917  975    5    3    7
## [5,]  203  207  183    0    0    1
## [6,]   10   11   20  222  408  313
## +++++++++++++++++++++
## &gt; subgenome: B 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]  222  254  255   50  119  124
## [2,]   27   36   27   49   50   36
## [3,]   52   47   46   11    7    6
## [4,]  750  579  583    4    8    2
## [5,]  318  382  410    0    1    1
## [6,]    0    0    0   17   16   17
## +++++++++++++++++++++
## &gt; subgenome: D 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]  203  212  199   36   92   90
## [2,]   50   51   57   37   53   40
## [3,]    5    6    4    3    0    2
## [4,]  473  441  475   12   16    4
## [5,]  456  454  427    6    1    3
## [6,]  129  127  146  497 1217 1189
## ---------------------</code></pre>
<p>Before performing the test,
we normalize the raw read counts using the TMM method <span class="citation">(Robinson and Oshlack 2010)</span>
to adjust for differences in library size.
However, if the expression data (<code>gexp</code>) has already been normalized (e.g. FPKM),
this step can be skipped.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="allohexaploid-t.-aestivum.html#cb37-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">norm_counts</span>(x)</span></code></pre></div>
<p>To visualize the distributions of homeolog expression ratios, use the <code>plot_HER_distr()</code> function.
By default, the expression ratio is calculated as the proportion contributed
by the first subgenome relative to all subgenomes.
In this dataset, the first subgenome corresponds to the A-subgenome,
as indicated by the first column of the mapping table (<code>mapping_table</code>).</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="allohexaploid-t.-aestivum.html#cb38-1" tabindex="-1"></a>distr_plots_A <span class="ot">&lt;-</span> <span class="fu">plot_HER_distr</span>(x)</span>
<span id="cb38-2"><a href="allohexaploid-t.-aestivum.html#cb38-2" tabindex="-1"></a><span class="fu">names</span>(distr_plots_A)</span></code></pre></div>
<pre><code>## [1] &quot;apex&quot; &quot;leaf&quot;</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="allohexaploid-t.-aestivum.html#cb40-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb40-2"><a href="allohexaploid-t.-aestivum.html#cb40-2" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb40-3"><a href="allohexaploid-t.-aestivum.html#cb40-3" tabindex="-1"></a></span>
<span id="cb40-4"><a href="allohexaploid-t.-aestivum.html#cb40-4" tabindex="-1"></a><span class="fu">grid.arrange</span>(distr_plots_A[[<span class="st">&quot;apex&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;apex on A subgenome&quot;</span>),</span>
<span id="cb40-5"><a href="allohexaploid-t.-aestivum.html#cb40-5" tabindex="-1"></a>             distr_plots_A[[<span class="st">&quot;leaf&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;leaf on A subgenome&quot;</span>),</span>
<span id="cb40-6"><a href="allohexaploid-t.-aestivum.html#cb40-6" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-wheat-viz-distr"></span>
<img src="_main_files/figure-html/tt-wheat-viz-distr-1.png" alt="Distribution of homeolog expression ratios between shoot apex and leaf in the A-subgenome of wheat." width="672" />
<p class="caption">
Figure 3.5: Distribution of homeolog expression ratios between shoot apex and leaf in the A-subgenome of wheat.
</p>
</div>
<p>To calculate expression ratios from the second or third subgenome,
set the <code>base</code> option accordingly.
For example, to calculate the ratios for the second and third subgenomes
(i.e., B-subgenome and D-subgenome), set <code>base = 2</code> and <code>base = 3</code>, respectively.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="allohexaploid-t.-aestivum.html#cb41-1" tabindex="-1"></a>distr_plots_B <span class="ot">&lt;-</span> <span class="fu">plot_HER_distr</span>(x, <span class="at">base =</span> <span class="dv">2</span>)</span>
<span id="cb41-2"><a href="allohexaploid-t.-aestivum.html#cb41-2" tabindex="-1"></a>distr_plots_D <span class="ot">&lt;-</span> <span class="fu">plot_HER_distr</span>(x, <span class="at">base =</span> <span class="dv">3</span>)</span>
<span id="cb41-3"><a href="allohexaploid-t.-aestivum.html#cb41-3" tabindex="-1"></a></span>
<span id="cb41-4"><a href="allohexaploid-t.-aestivum.html#cb41-4" tabindex="-1"></a><span class="fu">grid.arrange</span>(distr_plots_A[[<span class="st">&quot;apex&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;apex on A&quot;</span>), distr_plots_A[[<span class="st">&quot;leaf&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;leaf on A&quot;</span>),</span>
<span id="cb41-5"><a href="allohexaploid-t.-aestivum.html#cb41-5" tabindex="-1"></a>             distr_plots_B[[<span class="st">&quot;apex&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;apex on B&quot;</span>), distr_plots_B[[<span class="st">&quot;leaf&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;leaf on B&quot;</span>),</span>
<span id="cb41-6"><a href="allohexaploid-t.-aestivum.html#cb41-6" tabindex="-1"></a>             distr_plots_D[[<span class="st">&quot;apex&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;apex on D&quot;</span>), distr_plots_D[[<span class="st">&quot;leaf&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;leaf on D&quot;</span>),</span>
<span id="cb41-7"><a href="allohexaploid-t.-aestivum.html#cb41-7" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-wheat-viz-distrs"></span>
<img src="_main_files/figure-html/tt-wheat-viz-distrs-1.png" alt="Distribution of homeolog expression ratios between shoot apex and leaf across all subgenomes." width="672" />
<p class="caption">
Figure 3.6: Distribution of homeolog expression ratios between shoot apex and leaf across all subgenomes.
</p>
</div>
<p>Histogram peaks at one-third for all subgenomes indicate
that most homeologs in wheat are equally expressed from the three subgenomes.</p>
</div>
<div id="statistical-test" class="section level3 hasAnchor" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Statistical Test<a href="allohexaploid-t.-aestivum.html#statistical-test" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>After visually inspecting the distributions of homeolog expression ratios,
use the <code>hobit()</code> function to identify homeologs
showing significant changes in expression ratios between shoot apex and leaf tissues.
This step takes approximately five minutes using eight threads and may vary depending on hardware performance.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="allohexaploid-t.-aestivum.html#cb42-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x)</span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="allohexaploid-t.-aestivum.html#cb43-1" tabindex="-1"></a><span class="fu">head</span>(x_output)</span></code></pre></div>
<pre><code>##                 gene       pvalue       qvalue   raw_pvalue   raw_qvalue      Dmax    ORmax theta__A__apex theta__B__apex theta__D__apex theta__A__leaf theta__B__leaf theta__D__leaf theta0__A   theta0__B theta0__D logLik_H1  logLik_H0
## 1 TraesCS1A02G009000 8.696609e-03 2.643346e-02 1.133099e-03 3.444069e-03 0.1448480 1.815553      0.3587875    0.348272500      0.2918025     0.50432450     0.27980250      0.2099645 0.4319615 0.314246000 0.2513570 -86.71820  -93.45285
## 2 TraesCS1A02G011200 3.527622e-01 5.010827e-01 2.254854e-01 3.202917e-01 0.1320795 1.811967      0.2266725    0.276752000      0.4921475     0.18890200     0.40919000      0.3920710 0.2091740 0.344837000 0.4431745 -64.51790  -65.83085
## 3 TraesCS1A02G033800 5.438784e-02 1.155177e-01 1.557326e-02 3.308250e-02 0.1193433 3.906846      0.1647610    0.760828000      0.0714798     0.04790415     0.73972300      0.1923665 0.1114145 0.746659500 0.1338610 -39.94355  -43.89195
## 4 TraesCS1A02G038400 3.349351e-16 1.594929e-14 7.554204e-23 3.597240e-21 0.2787295 3.492158      0.4664375    0.308372000      0.2247815     0.24561800     0.22612250      0.5042635 0.3555395 0.268432000 0.3635740 -73.46170 -124.26850
## 5 TraesCS1A02G042500 2.729579e-09 4.402547e-08 5.724942e-13 9.233778e-12 0.2430330 2.744239      0.1947140    0.365449500      0.4387245     0.10636100     0.17506650      0.6810465 0.1501335 0.269729000 0.5585960 -56.57470  -84.81420
## 6 TraesCS1A02G045800 5.965812e-04 2.560434e-03 2.459053e-05 1.055388e-04 0.1645745 6.926095      0.0920326    0.002089305      0.9048555     0.24607950     0.01400095      0.7393605 0.1695830 0.008482395 0.8211180 -72.41100  -83.04700</code></pre>
<p>The output can be sorted in ascending order of <em>p</em>-values using the following code.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="allohexaploid-t.-aestivum.html#cb45-1" tabindex="-1"></a><span class="fu">head</span>(x_output[<span class="fu">order</span>(x_output<span class="sc">$</span>pvalue), ])</span></code></pre></div>
<pre><code>##                   gene       pvalue       qvalue   raw_pvalue   raw_qvalue      Dmax     ORmax theta__A__apex theta__B__apex theta__D__apex theta__A__leaf theta__B__leaf theta__D__leaf theta0__A theta0__B theta0__D  logLik_H1 logLik_H0
## 891 TraesCS7A02G209600 7.897716e-35 7.003971e-32 1.778006e-49 1.778006e-46 0.7214820  71.60537      0.4919380    0.233291500    0.273683000     0.02099795     0.95637300     0.02070660 0.2575725 0.5938435 0.1482310  -69.54315 -181.6070
## 933 TraesCS7A02G313100 1.400794e-34 7.003971e-32 4.033650e-49 2.016825e-46 0.2912081  10.04127      0.3159835    0.340996000    0.342377500     0.44160250     0.04913975     0.50780350 0.3791785 0.1956815 0.4249365 -115.24450 -226.5475
## 445 TraesCS3A02G438700 5.612771e-30 1.870924e-27 1.532781e-42 5.109269e-40 0.5502635  12.44399      0.3421380    0.272885500    0.383991000     0.06935165     0.82311350     0.06942305 0.2069145 0.5480310 0.2282040  -47.18400 -143.2310
## 214 TraesCS2A02G224600 6.437637e-29 1.609409e-26 5.013273e-41 1.253318e-38 0.5859770  41.88611      0.3238605    0.348283500    0.326649500     0.91179700     0.06739785     0.01141005 0.6174870 0.2092690 0.1714410  -58.26365 -150.7845
## 864 TraesCS7A02G119900 3.335147e-25 6.670295e-23 1.023063e-35 2.046127e-33 0.7192710 522.09016      0.9944905    0.002314735    0.002227525     0.27478050     0.29658300     0.29490100 0.6343520 0.1500650 0.1494245  -17.67280  -98.6582
## 328 TraesCS3A02G105600 2.636676e-22 4.394460e-20 1.420887e-31 2.368145e-29 0.5242690  10.37861      0.3936030    0.243851500    0.361295000     0.08921855     0.76699050     0.09134945 0.2423015 0.5055850 0.2268060  -46.26470 -117.5220</code></pre>
<p>Next, we visualize changes in expression ratios of the first subgenome between the two tissues.
In the plot, homeologs with <em>q</em>-values less than 0.01 are highlighted.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="allohexaploid-t.-aestivum.html#cb47-1" tabindex="-1"></a>is_sig <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(x_output<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>, <span class="st">&quot;q&lt;0.01&quot;</span>, <span class="st">&quot;q&gt;=0.01&quot;</span>)</span>
<span id="cb47-2"><a href="allohexaploid-t.-aestivum.html#cb47-2" tabindex="-1"></a><span class="fu">plot_HER</span>(x, <span class="at">label =</span> is_sig)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-wheat-viz-HER-changes"></span>
<img src="_main_files/figure-html/tt-wheat-viz-HER-changes-1.png" alt="Changes in homeolog expression ratios of the A-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes." width="672" />
<p class="caption">
Figure 3.7: Changes in homeolog expression ratios of the A-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes.
</p>
</div>
<p>To change the subgenome used for calculating expression ratios,
set the <code>base</code> argument to the corresponding subgenome number.
For example, to use the second subgenome, run the following.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="allohexaploid-t.-aestivum.html#cb48-1" tabindex="-1"></a><span class="fu">plot_HER</span>(x, <span class="at">base =</span> <span class="dv">2</span>, <span class="at">label =</span> is_sig)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-wheat-viz-HER-changes-base-2"></span>
<img src="_main_files/figure-html/tt-wheat-viz-HER-changes-base-2-1.png" alt="Changes in homeolog expression ratios of the B-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes." width="672" />
<p class="caption">
Figure 3.8: Changes in homeolog expression ratios of the B-subgenome between shoot apex and leaf tissues. Each point represents the ratio in shoot apex (x-axis) and leaf (y-axis). Orange points indicate homeologs with significant changes, gray points indicate homeologs without significant changes.
</p>
</div>
<p>Note that expression ratios are calculated for the selected subgenome,
while highlighted homeologs are those that meet the significance threshold in any subgenome.
As a result, some homeologs may appear highlighted even if they show only minor changes in the plot,
because they may have significantly altered expression ratios in another subgenome.</p>
</div>
<div id="combining-subgenomes" class="section level3 hasAnchor" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> Combining Subgenomes<a href="allohexaploid-t.-aestivum.html#combining-subgenomes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Additionally, users who want to test expression ratios
between the combined A- and B-subgenomes versus the D-subgenome
can first merge the expression values of the A and B homeologs
using the <code>combine_hexp()</code> function.
After combining, the statistical test can be performed on the merged dataset.
Assuming the A- and B-subgenomes are stored as the first and second subgenomes
in the <code>x</code> object (as defined in the <code>mapping_table</code>),
set <code>subgenomes = c(1, 2)</code> to combine them.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="allohexaploid-t.-aestivum.html#cb49-1" tabindex="-1"></a>x_ABvsD <span class="ot">&lt;-</span> <span class="fu">combine_hexp</span>(x, <span class="at">subgenomes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb49-2"><a href="allohexaploid-t.-aestivum.html#cb49-2" tabindex="-1"></a>x_ABvsD</span></code></pre></div>
<pre><code>## # 2 subgenomes (D, A+B)
## # 1000 homeologs
## ---------------------
## Experiment Design:
##   group
## 1  apex
## 2  apex
## 3  apex
## 4  leaf
## 5  leaf
## 6  leaf
## ---------------------
## &gt; subgenome: D 
##            [,1]       [,2]       [,3]       [,4]        [,5]        [,6]
## [1,] 155.071321 176.087014 157.993806  53.138696  115.656439  119.284316
## [2,]  38.194907  42.360555  45.254507  54.614770   66.628166   53.015252
## [3,]   3.819491   4.983595   3.175755   4.428225    0.000000    2.650763
## [4,] 361.323817 366.294213 377.120893  17.712899   20.114163    5.301525
## [5,] 348.337548 377.092002 339.011834   8.856449    1.257135    3.976144
## [6,]  98.542859 105.486089 115.915053 733.609214 1529.933544 1575.878355
## +++++++++++++++++++++
## &gt; subgenome: A+B 
##             [,1]       [,2]       [,3]      [,4]       [,5]       [,6]
## [1,]  365.143307  414.46896  404.90875 247.98058 419.883158 429.423538
## [2,]   37.431008   47.34415   46.04845 107.75347  93.028005  63.618302
## [3,]   49.653379   47.34415   43.66663  16.23682  10.057082   7.952288
## [4,] 1329.182751 1242.57629 1236.95653  13.28467  13.828487  11.928432
## [5,]  397.990927  489.22288  470.80566   0.00000   1.257135   2.650763
## [6,]    7.638981    9.13659   15.87877 352.78190 533.025327 437.375826
## ---------------------</code></pre>
<p>HOBIT can then be applied to the combined dataset as below.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="allohexaploid-t.-aestivum.html#cb51-1" tabindex="-1"></a>x_ABvsD_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x_ABvsD)</span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="allohexaploid-t.-aestivum.html#cb52-1" tabindex="-1"></a><span class="fu">head</span>(x_ABvsD_output[<span class="fu">order</span>(x_ABvsD_output<span class="sc">$</span>pvalue), ])</span></code></pre></div>
<pre><code>##                   gene       pvalue       qvalue   raw_pvalue   raw_qvalue      Dmax      ORmax theta__D__apex theta__A.B__apex theta__D__leaf theta__A.B__leaf theta0__D theta0__A.B logLik_H1  logLik_H0
## 474 TraesCS4A02G029000 2.930972e-19 2.930972e-16 7.663774e-27 7.663774e-24 0.2991075   9.440408    0.355223500       0.64477650     0.05499090        0.9450090 0.2062760   0.7937240 -40.87040  -98.35210
## 175 TraesCS2A02G145200 1.571885e-18 7.859426e-16 8.381772e-26 4.190886e-23 0.5913355  15.328096    0.194102500       0.80589750     0.78636800        0.2136320 0.4903530   0.5096470 -34.03465  -88.93955
## 864 TraesCS7A02G119900 4.130254e-17 1.223706e-14 8.806216e-24 2.803903e-21 0.4996285 523.905472    0.002271615       0.99772850     0.50338550        0.4966145 0.2530800   0.7469200 -16.43945  -67.43060
## 4   TraesCS1A02G038400 4.894824e-17 1.223706e-14 1.121561e-23 2.803903e-21 0.3188465   4.105482    0.224767000       0.77523300     0.54355100        0.4564490 0.3855575   0.6144425 -50.74050 -100.94100
## 214 TraesCS2A02G224600 9.173710e-16 1.834742e-13 7.272521e-22 1.454504e-19 0.3115651  45.588657    0.327089000       0.67291100     0.01044525        0.9895550 0.1710100   0.8289900 -41.09275  -87.00505
## 327 TraesCS3A02G098900 1.762103e-15 2.936838e-13 1.841413e-21 3.069021e-19 0.4922086 390.350835    0.997224500       0.00277519     0.50218300        0.4978170 0.7496330   0.2503670 -14.63465  -60.19225</code></pre>
</div>
<div id="analysis-environment-1" class="section level3 hasAnchor" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Analysis Environment<a href="allohexaploid-t.-aestivum.html#analysis-environment-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="allohexaploid-t.-aestivum.html#cb54-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.0 (2025-04-11)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sequoia 15.6.1
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: Asia/Tokyo
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] pROC_1.18.5     UpSetR_1.4.0    gridExtra_2.3   ggplot2_3.5.2   hespresso_1.0.0
## 
## loaded via a namespace (and not attached):
##  [1] tensorA_0.36.2.1     sass_0.4.10          generics_0.1.4       lattice_0.22-7       digest_0.6.37        magrittr_2.0.3       evaluate_1.0.3       grid_4.5.0           RColorBrewer_1.1-3   bookdown_0.44        iterators_1.0.14     fastmap_1.2.0        plyr_1.8.9           foreach_1.5.2        doParallel_1.0.17    jsonlite_2.0.0       processx_3.8.6       backports_1.5.0      limma_3.64.1         ps_1.9.1             scales_1.4.0         codetools_0.2-20     jquerylib_0.1.4      abind_1.4-8          cli_3.6.5            rlang_1.1.6          cmdstanr_0.9.0       splines_4.5.0        withr_3.0.2          cachem_1.1.0         yaml_2.3.10          tools_4.5.0          parallel_4.5.0       checkmate_2.3.2      dplyr_1.1.4          locfit_1.5-9.12      vctrs_0.6.5          posterior_1.6.1      R6_2.6.1             lifecycle_1.0.4      edgeR_4.6.2          pkgconfig_2.0.3      progressr_0.15.1     pillar_1.10.2        bslib_0.9.0          gtable_0.3.6         Rcpp_1.1.0           glue_1.8.0          
## [49] statmod_1.5.0        tidyselect_1.2.1     xfun_0.52            tibble_3.2.1         rstudioapi_0.17.1    knitr_1.50           farver_2.1.2         htmltools_0.5.8.1    labeling_0.4.3       rmarkdown_2.29       compiler_4.5.0       distributional_0.5.0</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="allohexaploid-c.-flexuosa.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="allotriploid-c.-insueta.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  },
  "number_sections": true,
  "biblio-style": "apalike",
  "link-citations": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
