<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.3 Allotriploid C. insueta | hespresso ☕</title>
  <meta name="description" content="3.3 Allotriploid C. insueta | hespresso ☕" />
  <meta name="generator" content="bookdown 0.44 and GitBook 2.6.7" />

  <meta property="og:title" content="3.3 Allotriploid C. insueta | hespresso ☕" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.3 Allotriploid C. insueta | hespresso ☕" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="allohexaploid-t.-aestivum.html"/>
<link rel="next" href="functions.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="_style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>hespresso</a></li>
<li class="chapter" data-level="1" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>1</b> Installation</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a></li>
<li class="chapter" data-level="3" data-path="tutorials.html"><a href="tutorials.html"><i class="fa fa-check"></i><b>3</b> Tutorials</a>
<ul>
<li class="chapter" data-level="3.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html"><i class="fa fa-check"></i><b>3.1</b> Allohexaploid <em>C. flexuosa</em></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#introduction"><i class="fa fa-check"></i><b>3.1.1</b> Introduction</a></li>
<li class="chapter" data-level="3.1.2" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#data-preparation"><i class="fa fa-check"></i><b>3.1.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.1.3" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#hobit"><i class="fa fa-check"></i><b>3.1.3</b> HOBIT</a></li>
<li class="chapter" data-level="3.1.4" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#homeoroq"><i class="fa fa-check"></i><b>3.1.4</b> HomeoRoq</a></li>
<li class="chapter" data-level="3.1.5" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#analysis-environment"><i class="fa fa-check"></i><b>3.1.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html"><i class="fa fa-check"></i><b>3.2</b> Allohexaploid <em>T. aestivum</em></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#introduction-1"><i class="fa fa-check"></i><b>3.2.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#data-preparation-1"><i class="fa fa-check"></i><b>3.2.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.2.3" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#statistical-test"><i class="fa fa-check"></i><b>3.2.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.2.4" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#combining-subgenomes"><i class="fa fa-check"></i><b>3.2.4</b> Combining Subgenomes</a></li>
<li class="chapter" data-level="3.2.5" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#analysis-environment-1"><i class="fa fa-check"></i><b>3.2.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html"><i class="fa fa-check"></i><b>3.3</b> Allotriploid <em>C. insueta</em></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#introduction-2"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#data-preparation-2"><i class="fa fa-check"></i><b>3.3.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#statistical-test-1"><i class="fa fa-check"></i><b>3.3.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.3.4" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#analysis-environment-2"><i class="fa fa-check"></i><b>3.3.4</b> Analysis Environment</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4</b> Functions</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hobit-1.html"><a href="hobit-1.html"><i class="fa fa-check"></i><b>4.1</b> HOBIT</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="hobit-1.html"><a href="hobit-1.html#inputs"><i class="fa fa-check"></i><b>4.1.1</b> Inputs</a></li>
<li class="chapter" data-level="4.1.2" data-path="hobit-1.html"><a href="hobit-1.html#test"><i class="fa fa-check"></i><b>4.1.2</b> Test</a></li>
<li class="chapter" data-level="4.1.3" data-path="hobit-1.html"><a href="hobit-1.html#outputs"><i class="fa fa-check"></i><b>4.1.3</b> Outputs</a></li>
<li class="chapter" data-level="4.1.4" data-path="hobit-1.html"><a href="hobit-1.html#options"><i class="fa fa-check"></i><b>4.1.4</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html"><i class="fa fa-check"></i><b>4.2</b> HomeoRoq</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="homeoroq-1.html"><a href="homeoroq-1.html#inputs-1"><i class="fa fa-check"></i><b>4.2.1</b> Inputs</a></li>
<li class="chapter" data-level="4.2.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html#test-1"><i class="fa fa-check"></i><b>4.2.2</b> Test</a></li>
<li class="chapter" data-level="4.2.3" data-path="homeoroq-1.html"><a href="homeoroq-1.html#outputs-1"><i class="fa fa-check"></i><b>4.2.3</b> Outputs</a></li>
<li class="chapter" data-level="4.2.4" data-path="homeoroq-1.html"><a href="homeoroq-1.html#options-1"><i class="fa fa-check"></i><b>4.2.4</b> Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="evaluation-framework.html"><a href="evaluation-framework.html"><i class="fa fa-check"></i><b>5</b> Evaluation Framework</a>
<ul>
<li class="chapter" data-level="5.1" data-path="simulating-allotetraploid-data.html"><a href="simulating-allotetraploid-data.html"><i class="fa fa-check"></i><b>5.1</b> Simulating Allotetraploid Data</a></li>
<li class="chapter" data-level="5.2" data-path="simulating-allohexaploid-data.html"><a href="simulating-allohexaploid-data.html"><i class="fa fa-check"></i><b>5.2</b> Simulating Allohexaploid Data</a></li>
<li class="chapter" data-level="5.3" data-path="custom-data-simulation.html"><a href="custom-data-simulation.html"><i class="fa fa-check"></i><b>5.3</b> Custom Data Simulation</a></li>
<li class="chapter" data-level="5.4" data-path="evaluation-framework-1.html"><a href="evaluation-framework-1.html"><i class="fa fa-check"></i><b>5.4</b> Evaluation Framework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="citations.html"><a href="citations.html"><i class="fa fa-check"></i>Citations</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><em>hespresso</em> ☕</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="allotriploid-c.-insueta" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Allotriploid <em>C. insueta</em><a href="allotriploid-c.-insueta.html#allotriploid-c.-insueta" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="introduction-2" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Introduction<a href="allotriploid-c.-insueta.html#introduction-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Allotriploid <em>Cardamine insueta</em> (2n = 3x = 24, RRA) serves as an example of an uneven subgenome composition.
Two of its subgenomes are derived from <em>Cardamine rivularis</em> (2n = 2x = 16, RR),
a species inhabiting grassy areas away from riversides,
while the third subgenome comes from <em>Cardamine amara</em> (2n = 2x = 16, AA),
a species typically found on riversides or in streams <span class="citation">(Urbanska-Worytkiewicz and Landolt 1972, 1974)</span>.
RNA-Seq studies on <em>C. insueta</em> leaflets floating on water suggested that
<em>C. insueta</em> shifts homeolog expression ratios over time following submergence <span class="citation">(Sun et al. 2020)</span>.</p>
</div>
<div id="data-preparation-2" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Data Preparation<a href="allotriploid-c.-insueta.html#data-preparation-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here, we demonstrate how to detect homeologs with changing expression ratios across time points using HOBIT,
similar to an ANOVA approach.
We use a sample dataset containing 1,000 homeologs (<code>c_insueta.mini.txt.gz</code>) randomly selected from the original experiment,
although the full dataset is available in the <a href="https://github.com/bitdessin/hespresso/tree/main/docs">data</a> directory
for users who wish to analyze it.
This data span nine time points following leaf submergence and can be loaded using the following code.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="allotriploid-c.-insueta.html#cb56-1" tabindex="-1"></a>gexp <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_insueta.mini.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb56-2"><a href="allotriploid-c.-insueta.html#cb56-2" tabindex="-1"></a><span class="fu">head</span>(gexp)</span></code></pre></div>
<pre><code>##               T00 T02 T04 T08 T12 T24 T48 T72 T96
## CARHR000430_R 148  32 134  76 118  54 119 144  94
## CARHR000770_R  88 127  17   2  56  25  35  22  17
## CARHR000870_R 157  37 194  82 129  57 110  98  98
## CARHR001290_R  52  18  84  35  63  12  33  43  34
## CARHR001460_R 735 131 743 523 734 315 499 582 369
## CARHR001580_R 420  67 344 327 585 177 341 475 368</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="allotriploid-c.-insueta.html#cb58-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;T00&quot;</span>, <span class="st">&quot;T02&quot;</span>, <span class="st">&quot;T04&quot;</span>, <span class="st">&quot;T08&quot;</span>, <span class="st">&quot;T12&quot;</span>, <span class="st">&quot;T24&quot;</span>, <span class="st">&quot;T48&quot;</span>, <span class="st">&quot;T72&quot;</span>, <span class="st">&quot;T96&quot;</span>)</span></code></pre></div>
<p>Next, load the homeolog mapping table,
which links gene expression values to their corresponding homeolog pairs.
This file is tab-separated,
with the first and second columns representing genes from <em>C. rivularis</em> and <em>C. amara</em>, respectively.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="allotriploid-c.-insueta.html#cb59-1" tabindex="-1"></a>mapping_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_insueta.homeolog.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb59-2"><a href="allotriploid-c.-insueta.html#cb59-2" tabindex="-1"></a><span class="fu">head</span>(mapping_table)</span></code></pre></div>
<pre><code>##     C_rivularis       C_amara
## 1 CARHR000010_R CARHR000010_A
## 2 CARHR000060_R CARHR000060_A
## 3 CARHR000090_R CARHR000090_A
## 4 CARHR000110_R CARHR000110_A
## 5 CARHR000120_R CARHR000120_A
## 6 CARHR000130_R CARHR000130_A</code></pre>
<p>Use the <code>newExpMX()</code> function
to create a homeolog expression matrix from the gene expression data (<code>gexp</code>)
and mapping table (<code>mapping_table</code>).
The result is stored as an <code>ExpMX</code> class object.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="allotriploid-c.-insueta.html#cb61-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">newExpMX</span>(gexp, group, mapping_table)</span>
<span id="cb61-2"><a href="allotriploid-c.-insueta.html#cb61-2" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>## # 2 subgenomes (C_rivularis, C_amara)
## # 1000 homeologs
## ---------------------
## Experiment Design:
##   group
## 1   T00
## 2   T02
## 3   T04
## 4   T08
## 5   T12
## 6   T24
## 7   T48
## 8   T72
## 9   T96
## ---------------------
## &gt; subgenome: C_rivularis 
##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
## [1,]  148   32  134   76  118   54  119  144   94
## [2,]   88  127   17    2   56   25   35   22   17
## [3,]  157   37  194   82  129   57  110   98   98
## [4,]   52   18   84   35   63   12   33   43   34
## [5,]  735  131  743  523  734  315  499  582  369
## [6,]  420   67  344  327  585  177  341  475  368
## +++++++++++++++++++++
## &gt; subgenome: C_amara 
##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
## [1,]   45   15   42   34   61   34   47   44   38
## [2,]   30   64   10   21   54   17   24   15   11
## [3,]   98   38  107   45   61   34  108   75   37
## [4,]   20   13   54   25   26    9   23   18   12
## [5,]  416   87  409  331  457  196  322  393  321
## [6,]  165   19  115  147  264   67  127  160  119
## ---------------------</code></pre>
<p>Normalize the read counts using the TMM method <span class="citation">(Robinson and Oshlack 2010)</span>,
since the data has not been normalized.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="allotriploid-c.-insueta.html#cb63-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">norm_counts</span>(x)</span></code></pre></div>
<p>Users can visualize the distribution of homeolog expression ratios using the <code>plot_HER_distr()</code> function.
By default, the expression ratio is calculated as
the proportion contributed by the first subgenome relative to all subgenomes,
where the first subgenome corresponds to <em>C. rivularis</em>,
as listed in the first column of the mapping table (<code>mapping_table</code>).
The <code>plot_HER_distr()</code> function generates a histogram for each time point.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="allotriploid-c.-insueta.html#cb64-1" tabindex="-1"></a>distr_plots <span class="ot">&lt;-</span> <span class="fu">plot_HER_distr</span>(x)</span>
<span id="cb64-2"><a href="allotriploid-c.-insueta.html#cb64-2" tabindex="-1"></a></span>
<span id="cb64-3"><a href="allotriploid-c.-insueta.html#cb64-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb64-4"><a href="allotriploid-c.-insueta.html#cb64-4" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb64-5"><a href="allotriploid-c.-insueta.html#cb64-5" tabindex="-1"></a>distr_plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(distr_plots), <span class="cf">function</span>(p) {</span>
<span id="cb64-6"><a href="allotriploid-c.-insueta.html#cb64-6" tabindex="-1"></a>    distr_plots[[p]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="fu">names</span>(distr_plots)[p])</span>
<span id="cb64-7"><a href="allotriploid-c.-insueta.html#cb64-7" tabindex="-1"></a>})</span>
<span id="cb64-8"><a href="allotriploid-c.-insueta.html#cb64-8" tabindex="-1"></a><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> distr_plots, <span class="at">ncol =</span> <span class="dv">3</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-cins-viz-HER-distr"></span>
<img src="_main_files/figure-html/tt-cins-viz-HER-distr-1.png" alt="Distribution of homeolog expression ratios in _C. insueta_ across nine time points following leaf submergence in water." width="672" />
<p class="caption">
Figure 3.9: Distribution of homeolog expression ratios in <em>C. insueta</em> across nine time points following leaf submergence in water.
</p>
</div>
<p>These visualizations indicate that most homeologs in <em>C. insueta</em> are expressed
at a 2:1 ratio between the <em>C. rivularis</em> and <em>C. amara</em> subgenomes,
as shown by histogram peaks centered around two-thirds.</p>
</div>
<div id="statistical-test-1" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Statistical Test<a href="allotriploid-c.-insueta.html#statistical-test-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Next, apply the <code>hobit()</code> function
to detect homeologs with differential expression ratios across the nine time points following leaf submergence.
Since this dataset contains only one replicate per condition,
set <code>no_replicate = TRUE</code> to avoid errors arising from dispersion estimation,
which requires multiple replicates.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="allotriploid-c.-insueta.html#cb65-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">no_replicate =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The test takes approximately five minutes to analyze 1,000 homeolog pairs using eight threads,
although runtime may vary depending on hardware performance.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="allotriploid-c.-insueta.html#cb66-1" tabindex="-1"></a><span class="fu">head</span>(x_output)</span></code></pre></div>
<pre><code>##            gene    pvalue qvalue raw_pvalue raw_qvalue      Dmax    ORmax theta__C_rivularis__T00 theta__C_amara__T00 theta__C_rivularis__T02 theta__C_amara__T02 theta__C_rivularis__T04 theta__C_amara__T04 theta__C_rivularis__T08 theta__C_amara__T08 theta__C_rivularis__T12 theta__C_amara__T12 theta__C_rivularis__T24 theta__C_amara__T24 theta__C_rivularis__T48 theta__C_amara__T48 theta__C_rivularis__T72 theta__C_amara__T72 theta__C_rivularis__T96 theta__C_amara__T96 theta0__C_rivularis theta0__C_amara  logLik_H1  logLik_H0
## 1 CARHR000430_R 1.0000000      1  1.0000000          1 0.1438845 2.013606               0.7533710           0.2466290               0.6699725           0.3300275               0.7497035           0.2502965               0.6787245           0.3212755               0.6467840           0.3532160               0.6046150           0.3953850               0.7031835           0.2968165               0.7531800           0.2468200               0.7045160           0.2954840           0.6869845       0.3130155  -80.92400  -79.08280
## 2 CARHR000770_R 0.7508003      1  0.7508003          1 0.3470615 6.876720               0.7165820           0.2834180               0.6334590           0.3665410               0.6446150           0.3553850               0.2751455           0.7248545               0.5714540           0.4285460               0.6363635           0.3636365               0.6374075           0.3625925               0.6397045           0.3602955               0.6432505           0.3567495           0.5811455       0.4188545  -90.18315  -92.87335
## 3 CARHR000870_R 1.0000000      1  1.0000000          1 0.2229850 2.654365               0.6072565           0.3927435               0.4928865           0.5071135               0.6365870           0.3634130               0.6339005           0.3660995               0.6735320           0.3264680               0.6207140           0.3792860               0.5024625           0.4975375               0.5621570           0.4378430               0.7158015           0.2841985           0.5992305       0.4007695  -87.31455  -86.08080
## 4 CARHR001290_R 1.0000000      1  1.0000000          1 0.1566040 2.094365               0.6936415           0.3063585               0.5618700           0.4381300               0.5884945           0.4115055               0.5595320           0.4404680               0.6767025           0.3232975               0.5444725           0.4555275               0.5622140           0.4377860               0.6802175           0.3197825               0.7084995           0.2915005           0.6056650       0.3943350  -75.51685  -74.23905
## 5 CARHR001460_R 1.0000000      1  1.0000000          1 0.1027260 1.549844               0.6275635           0.3724365               0.5940575           0.4059425               0.6332530           0.3667470               0.6044705           0.3955295               0.6048025           0.3951975               0.6086210           0.3913790               0.5993465           0.4006535               0.5903205           0.4096795               0.5264095           0.4735905           0.5944350       0.4055650 -114.15700 -111.36700
## 6 CARHR001580_R 1.0000000      1  1.0000000          1 0.0814270 1.526471               0.6900865           0.3099135               0.7413470           0.2586530               0.7172220           0.2827780               0.6608300           0.3391700               0.6532005           0.3467995               0.6957720           0.3042280               0.6996390           0.3003610               0.7128460           0.2871540               0.7250950           0.2749050           0.6830990       0.3169010 -106.90800 -104.81250</code></pre>
</div>
<div id="analysis-environment-2" class="section level3 hasAnchor" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Analysis Environment<a href="allotriploid-c.-insueta.html#analysis-environment-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="allotriploid-c.-insueta.html#cb68-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.0 (2025-04-11)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sequoia 15.6.1
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: Asia/Tokyo
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] pROC_1.18.5     UpSetR_1.4.0    gridExtra_2.3   ggplot2_3.5.2   hespresso_1.0.0
## 
## loaded via a namespace (and not attached):
##  [1] tensorA_0.36.2.1     sass_0.4.10          generics_0.1.4       lattice_0.22-7       digest_0.6.37        magrittr_2.0.3       evaluate_1.0.3       grid_4.5.0           RColorBrewer_1.1-3   bookdown_0.44        iterators_1.0.14     fastmap_1.2.0        plyr_1.8.9           foreach_1.5.2        doParallel_1.0.17    jsonlite_2.0.0       processx_3.8.6       backports_1.5.0      limma_3.64.1         ps_1.9.1             scales_1.4.0         codetools_0.2-20     jquerylib_0.1.4      abind_1.4-8          cli_3.6.5            rlang_1.1.6          cmdstanr_0.9.0       splines_4.5.0        withr_3.0.2          cachem_1.1.0         yaml_2.3.10          tools_4.5.0          parallel_4.5.0       checkmate_2.3.2      dplyr_1.1.4          locfit_1.5-9.12      vctrs_0.6.5          posterior_1.6.1      R6_2.6.1             lifecycle_1.0.4      edgeR_4.6.2          pkgconfig_2.0.3      progressr_0.15.1     pillar_1.10.2        bslib_0.9.0          gtable_0.3.6         Rcpp_1.1.0           glue_1.8.0          
## [49] statmod_1.5.0        tidyselect_1.2.1     xfun_0.52            tibble_3.2.1         rstudioapi_0.17.1    knitr_1.50           farver_2.1.2         htmltools_0.5.8.1    labeling_0.4.3       rmarkdown_2.29       compiler_4.5.0       distributional_0.5.0</code></pre>

</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="allohexaploid-t.-aestivum.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="functions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  },
  "number_sections": true,
  "biblio-style": "apalike",
  "link-citations": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
