<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.1 Allohexaploid C. flexuosa | hespresso ☕</title>
  <meta name="description" content="3.1 Allohexaploid C. flexuosa | hespresso ☕" />
  <meta name="generator" content="bookdown 0.44 and GitBook 2.6.7" />

  <meta property="og:title" content="3.1 Allohexaploid C. flexuosa | hespresso ☕" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.1 Allohexaploid C. flexuosa | hespresso ☕" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="tutorials.html"/>
<link rel="next" href="allohexaploid-t.-aestivum.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="_style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>hespresso</a></li>
<li class="chapter" data-level="1" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>1</b> Installation</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a></li>
<li class="chapter" data-level="3" data-path="tutorials.html"><a href="tutorials.html"><i class="fa fa-check"></i><b>3</b> Tutorials</a>
<ul>
<li class="chapter" data-level="3.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html"><i class="fa fa-check"></i><b>3.1</b> Allohexaploid <em>C. flexuosa</em></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#introduction"><i class="fa fa-check"></i><b>3.1.1</b> Introduction</a></li>
<li class="chapter" data-level="3.1.2" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#data-preparation"><i class="fa fa-check"></i><b>3.1.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.1.3" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#hobit"><i class="fa fa-check"></i><b>3.1.3</b> HOBIT</a></li>
<li class="chapter" data-level="3.1.4" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#homeoroq"><i class="fa fa-check"></i><b>3.1.4</b> HomeoRoq</a></li>
<li class="chapter" data-level="3.1.5" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#analysis-environment"><i class="fa fa-check"></i><b>3.1.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html"><i class="fa fa-check"></i><b>3.2</b> Allohexaploid <em>T. aestivum</em></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#introduction-1"><i class="fa fa-check"></i><b>3.2.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#data-preparation-1"><i class="fa fa-check"></i><b>3.2.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.2.3" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#statistical-test"><i class="fa fa-check"></i><b>3.2.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.2.4" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#combining-subgenomes"><i class="fa fa-check"></i><b>3.2.4</b> Combining Subgenomes</a></li>
<li class="chapter" data-level="3.2.5" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#analysis-environment-1"><i class="fa fa-check"></i><b>3.2.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html"><i class="fa fa-check"></i><b>3.3</b> Allotriploid <em>C. insueta</em></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#introduction-2"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#data-preparation-2"><i class="fa fa-check"></i><b>3.3.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#statistical-test-1"><i class="fa fa-check"></i><b>3.3.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.3.4" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#analysis-environment-2"><i class="fa fa-check"></i><b>3.3.4</b> Analysis Environment</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4</b> Functions</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hobit-1.html"><a href="hobit-1.html"><i class="fa fa-check"></i><b>4.1</b> HOBIT</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="hobit-1.html"><a href="hobit-1.html#inputs"><i class="fa fa-check"></i><b>4.1.1</b> Inputs</a></li>
<li class="chapter" data-level="4.1.2" data-path="hobit-1.html"><a href="hobit-1.html#test"><i class="fa fa-check"></i><b>4.1.2</b> Test</a></li>
<li class="chapter" data-level="4.1.3" data-path="hobit-1.html"><a href="hobit-1.html#outputs"><i class="fa fa-check"></i><b>4.1.3</b> Outputs</a></li>
<li class="chapter" data-level="4.1.4" data-path="hobit-1.html"><a href="hobit-1.html#options"><i class="fa fa-check"></i><b>4.1.4</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html"><i class="fa fa-check"></i><b>4.2</b> HomeoRoq</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="homeoroq-1.html"><a href="homeoroq-1.html#inputs-1"><i class="fa fa-check"></i><b>4.2.1</b> Inputs</a></li>
<li class="chapter" data-level="4.2.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html#test-1"><i class="fa fa-check"></i><b>4.2.2</b> Test</a></li>
<li class="chapter" data-level="4.2.3" data-path="homeoroq-1.html"><a href="homeoroq-1.html#outputs-1"><i class="fa fa-check"></i><b>4.2.3</b> Outputs</a></li>
<li class="chapter" data-level="4.2.4" data-path="homeoroq-1.html"><a href="homeoroq-1.html#options-1"><i class="fa fa-check"></i><b>4.2.4</b> Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="evaluation-framework.html"><a href="evaluation-framework.html"><i class="fa fa-check"></i><b>5</b> Evaluation Framework</a>
<ul>
<li class="chapter" data-level="5.1" data-path="simulating-allotetraploid-data.html"><a href="simulating-allotetraploid-data.html"><i class="fa fa-check"></i><b>5.1</b> Simulating Allotetraploid Data</a></li>
<li class="chapter" data-level="5.2" data-path="simulating-allohexaploid-data.html"><a href="simulating-allohexaploid-data.html"><i class="fa fa-check"></i><b>5.2</b> Simulating Allohexaploid Data</a></li>
<li class="chapter" data-level="5.3" data-path="custom-data-simulation.html"><a href="custom-data-simulation.html"><i class="fa fa-check"></i><b>5.3</b> Custom Data Simulation</a></li>
<li class="chapter" data-level="5.4" data-path="evaluation-framework-1.html"><a href="evaluation-framework-1.html"><i class="fa fa-check"></i><b>5.4</b> Evaluation Framework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="citations.html"><a href="citations.html"><i class="fa fa-check"></i>Citations</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><em>hespresso</em> ☕</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="allohexaploid-c.-flexuosa" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Allohexaploid <em>C. flexuosa</em><a href="allohexaploid-c.-flexuosa.html#allohexaploid-c.-flexuosa" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="introduction" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Introduction<a href="allohexaploid-c.-flexuosa.html#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><em>Cardamine flexuosa</em> (2<em>n</em> = 4<em>x</em> = 32, HHAA), wavy bittercress,
is an allopolyploid species that originated from two diploid progenitors,
<em>Cardamine hirsuta</em> (2<em>n</em> = 2<em>x</em> = 16, HH) and <em>Cardamine amara</em> (2<em>n</em> = 2<em>x</em> = 16, AA) <span class="citation">(Mandáková et al. 2014)</span>.
While <em>C. hirsuta</em> is typically adapted to dry habitats such as roadsides,
<em>C. amara</em> inhabits wet environments such as riversides or running streams.
The allopolyploid <em>C. flexuosa</em> thrives across a wide range of ecological habitats.</p>
<p>Akiyama et al. conducted an RNA-Seq experiment using leaf samples of <em>C. flexuosa</em>
collected from wet and dry habitats across three different days <span class="citation">(Akiyama et al. 2021)</span>.
Each habitat–date combination included two or three biological replicates.
Homeolog expression levels were quantified using read counts
processed through the HomeoRoq sorting pipeline <span class="citation">(Akama et al. 2014)</span>.
Analysis of homeolog expression ratios revealed that
a small percentage of homeologs in <em>C. flexuosa</em> exhibit shifts
in expression depending on environmental differences, wet and dry.</p>
</div>
<div id="data-preparation" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Data Preparation<a href="allohexaploid-c.-flexuosa.html#data-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this example, we demonstrate how to use HOBIT and HomeoRoq to analyze the <em>C. flexuosa</em> dataset.
To reduce computation time, we focus on samples collected on May 16, 2013,
and randomly select 1,000 homeolog pairs (<code>c_flexuosa.0516.mini.txt.gz</code>).
Users can, however, analyze the full dataset or samples from other dates.
These datasets are available in the <a href="https://github.com/bitdessin/hespresso/tree/main/docs/data">data</a> directory,
and the same workflow can be applied.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="allohexaploid-c.-flexuosa.html#cb6-1" tabindex="-1"></a>gexp <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.0516.mini.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb6-2"><a href="allohexaploid-c.-flexuosa.html#cb6-2" tabindex="-1"></a><span class="fu">head</span>(gexp)</span></code></pre></div>
<pre><code>##               wet_1 wet_2 wet_3 dry_1 dry_2 dry_3
## CARHR000350_H    21     8    17    19    10    14
## CARHR000580_H    24    21    14    26    11    19
## CARHR001360_H   440   264   364   394   272   248
## CARHR001380_H    36   109    48    52    52    32
## CARHR001400_H    10    11    59    30    23    27
## CARHR001480_H    27    17    49    40    25    15</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="allohexaploid-c.-flexuosa.html#cb8-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>)</span></code></pre></div>
<p>Next, load the homeolog mapping table,
which links gene expression values to their corresponding homeologs.
This table is a tab-separated file in which the first and second columns
represent gene names from <em>C. hirsuta</em> and <em>C. amara</em>, respectively.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="allohexaploid-c.-flexuosa.html#cb9-1" tabindex="-1"></a>mapping_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.homeolog.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb9-2"><a href="allohexaploid-c.-flexuosa.html#cb9-2" tabindex="-1"></a><span class="fu">head</span>(mapping_table)</span></code></pre></div>
<pre><code>##       C_hirsuta       C_amara
## 1 CARHR000010_H CARHR000010_A
## 2 CARHR000060_H CARHR000060_A
## 3 CARHR000070_H CARHR000070_A
## 4 CARHR000090_H CARHR000090_A
## 5 CARHR000110_H CARHR000110_A
## 6 CARHR000120_H CARHR000120_A</code></pre>
<p>We then use the <code>newExpMX()</code> function
to organize the gene expression matrix (<code>gexp</code>) into a homeolog expression matrix
using the mapping table (<code>mapping_table</code>) and store the result in an <code>ExpMX</code> class object.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="allohexaploid-c.-flexuosa.html#cb11-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">newExpMX</span>(gexp, group, mapping_table)</span>
<span id="cb11-2"><a href="allohexaploid-c.-flexuosa.html#cb11-2" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>## # 2 subgenomes (C_hirsuta, C_amara)
## # 1000 homeologs
## ---------------------
## Experiment Design:
##   group
## 1   wet
## 2   wet
## 3   wet
## 4   dry
## 5   dry
## 6   dry
## ---------------------
## &gt; subgenome: C_hirsuta 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]   21    8   17   19   10   14
## [2,]   24   21   14   26   11   19
## [3,]  440  264  364  394  272  248
## [4,]   36  109   48   52   52   32
## [5,]   10   11   59   30   23   27
## [6,]   27   17   49   40   25   15
## +++++++++++++++++++++
## &gt; subgenome: C_amara 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]  152  168  183  151   92   75
## [2,]   54   44   49   46   22   27
## [3,]  344  181  242  257  203  183
## [4,]   73  150   49   43   37   32
## [5,]    4    0    2    2    6    1
## [6,]   30   23  124   48   32   28
## ---------------------</code></pre>
<p>Before performing the test,
we normalize the raw read counts using the TMM method <span class="citation">(Robinson and Oshlack 2010)</span>
to adjust for differences in library size.
However, if the expression data (<code>gexp</code>) has already been normalized (e.g. FPKM),
this step can be skipped.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="allohexaploid-c.-flexuosa.html#cb13-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">norm_counts</span>(x)</span></code></pre></div>
<p>To visualize the distributions of homeolog expression ratios,
use the <code>plot_HER_distr()</code> function.
This function returns a list of histograms, one for each condition,
showing the distribution of homeolog expression ratios.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="allohexaploid-c.-flexuosa.html#cb14-1" tabindex="-1"></a>distr_plots <span class="ot">&lt;-</span> <span class="fu">plot_HER_distr</span>(x)</span>
<span id="cb14-2"><a href="allohexaploid-c.-flexuosa.html#cb14-2" tabindex="-1"></a><span class="fu">names</span>(distr_plots)</span></code></pre></div>
<pre><code>## [1] &quot;wet&quot; &quot;dry&quot;</code></pre>
<p>By default, the expression ratio is calculated as the proportion contributed
by the first subgenome relative to all subgenomes.
In this dataset, the first subgenome is derived from <em>C. hirsuta</em>,
as indicated by the first column in the mapping table (<code>mapping_table</code>).</p>
<p>To display the distribution of homeolog expression ratios for each experimental group,
run the following code.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="allohexaploid-c.-flexuosa.html#cb16-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb16-2"><a href="allohexaploid-c.-flexuosa.html#cb16-2" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb16-3"><a href="allohexaploid-c.-flexuosa.html#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="allohexaploid-c.-flexuosa.html#cb16-4" tabindex="-1"></a><span class="fu">grid.arrange</span>(distr_plots[[<span class="st">&quot;wet&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;wet&quot;</span>),</span>
<span id="cb16-5"><a href="allohexaploid-c.-flexuosa.html#cb16-5" tabindex="-1"></a>             distr_plots[[<span class="st">&quot;dry&quot;</span>]] <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;dry&quot;</span>),</span>
<span id="cb16-6"><a href="allohexaploid-c.-flexuosa.html#cb16-6" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-cflex-viz-distr-show"></span>
<img src="_main_files/figure-html/tt-cflex-viz-distr-show-1.png" alt="Distribution of homeolog expression ratios in _C. flexuosa_ under wet and dry habitats." width="672" />
<p class="caption">
Figure 3.1: Distribution of homeolog expression ratios in <em>C. flexuosa</em> under wet and dry habitats.
</p>
</div>
<p>After visually inspecting the distributions,
if no irregularities are observed, users can proceed to apply the statistical test.</p>
</div>
<div id="hobit" class="section level3 hasAnchor" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> HOBIT<a href="allohexaploid-c.-flexuosa.html#hobit" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>HOBIT can be executed using the <code>hobit()</code> function, which performs a statistical test
to detect homeologs with shifts in expression ratios between wet and dry habitats.
Analyzing 1,000 homeolog pairs with eight threads takes approximately five minutes.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="allohexaploid-c.-flexuosa.html#cb17-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x)</span></code></pre></div>
<p>The order of the test output matches that of the input.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="allohexaploid-c.-flexuosa.html#cb18-1" tabindex="-1"></a><span class="fu">head</span>(x_output)</span></code></pre></div>
<pre><code>##            gene    pvalue qvalue raw_pvalue raw_qvalue      Dmax    ORmax theta__C_hirsuta__dry theta__C_amara__dry theta__C_hirsuta__wet theta__C_amara__wet theta0__C_hirsuta theta0__C_amara logLik_H1 logLik_H0
## 1 CARHR000350_H 0.3253319      1  0.2396078          1 0.0448685 1.663141             0.1253715           0.8746285            0.07960025          0.92039950         0.1036990      0.89630100 -47.05505 -47.70480
## 2 CARHR000580_H 0.6133947      1  0.5457975          1 0.0821525 1.465043             0.3767205           0.6232795            0.29091800          0.70908200         0.3353605      0.66463950 -42.30920 -42.35770
## 3 CARHR001360_H 1.0000000      1  1.0000000          1 0.0004880 1.001988             0.5810760           0.4189240            0.58084650          0.41915350         0.5802035      0.41979650 -67.25310 -66.94940
## 4 CARHR001380_H 0.3431881      1  0.2570826          1 0.1320365 1.708895             0.5500030           0.4499970            0.41447950          0.58552050         0.4831110      0.51688900 -67.96410 -68.75220
## 5 CARHR001400_H 0.6481960      1  0.5853912          1 0.0459866 1.846449             0.8834420           0.1165580            0.93359950          0.06640005         0.9043115      0.09568855 -41.85770 -41.90815
## 6 CARHR001480_H 1.0000000      1  1.0000000          1 0.0537910 1.261780             0.4093955           0.5906045            0.35391850          0.64608150         0.3845520      0.61544800 -54.03790 -53.88225</code></pre>
<p>To rank the output in descending order of <em>p</em>-values, use the following code.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="allohexaploid-c.-flexuosa.html#cb20-1" tabindex="-1"></a><span class="fu">head</span>(x_output[<span class="fu">order</span>(x_output<span class="sc">$</span>pvalue), ])</span></code></pre></div>
<pre><code>##              gene       pvalue       qvalue   raw_pvalue   raw_qvalue      Dmax      ORmax theta__C_hirsuta__dry theta__C_amara__dry theta__C_hirsuta__wet theta__C_amara__wet theta0__C_hirsuta theta0__C_amara  logLik_H1  logLik_H0
## 980 CARHR289110_H 3.127442e-08 1.573060e-05 3.671593e-11 1.851358e-08 0.4635909 128.904980             0.5266430           0.4733570             0.9930045          0.00699555         0.7576215       0.2423785  -31.33400  -53.03600
## 537 CARHR155190_H 3.146119e-08 1.573060e-05 3.702717e-11 1.851358e-08 0.4449000   7.874459             0.3839395           0.6160605             0.8334040          0.16659600         0.6085220       0.3914780 -135.15100 -157.04800
## 12  CARHR002710_H 2.349679e-07 6.145125e-05 6.388189e-10 1.702323e-07 0.6906636 206.664157             0.0122882           0.9877120             0.7176940          0.28230600         0.3715680       0.6284320  -75.89975  -94.52055
## 135 CARHR032730_H 2.458050e-07 6.145125e-05 6.809291e-10 1.702323e-07 0.2246896  29.324336             0.7648575           0.2351425             0.9895090          0.01049100         0.8769085       0.1230915  -46.14340  -65.39720
## 481 CARHR139340_H 2.714965e-06 5.429929e-04 2.034879e-08 4.069758e-06 0.3498725   4.443672             0.6189910           0.3810090             0.2654020          0.73459800         0.4437415       0.5562585 -183.45350 -199.11850
## 346 CARHR099310_H 3.022336e-05 4.437007e-03 6.104695e-07 9.062792e-05 0.3087950   4.443908             0.1754890           0.8245110             0.4880245          0.51197550         0.3332475       0.6667525  -56.89240  -69.17890</code></pre>
<p>To visualize homeologs with shifts in expression ratios between wet and dry habitats,
use the <code>plot_HER()</code> function.
This function generates a scatter plot comparing the expression ratios
calculated for the wet and dry groups.
By default, the expression ratio is defined as the proportion contributed
by the first subgenome (i.e., the subgenome derived from <em>C. hirsuta</em>)
relative to the total expression across all subgenomes.
In the example below, homeologs with <em>q</em>-values less than 0.01 are highlighted.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="allohexaploid-c.-flexuosa.html#cb22-1" tabindex="-1"></a>is_sig <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(x_output<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>, <span class="st">&quot;q&lt;0.01&quot;</span>, <span class="st">&quot;q&gt;=0.01&quot;</span>)</span>
<span id="cb22-2"><a href="allohexaploid-c.-flexuosa.html#cb22-2" tabindex="-1"></a><span class="fu">plot_HER</span>(x, <span class="at">label =</span> is_sig)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-cflex-viz-highlight"></span>
<img src="_main_files/figure-html/tt-cflex-viz-highlight-1.png" alt="Changes in homeolog expression ratios between wet (x-axis) and dry (y-axis) habitats. Orange points indicate homeologs with significant shifts detected by HOBIT, while gray points represent homeologs without significant changes." width="672" />
<p class="caption">
Figure 3.2: Changes in homeolog expression ratios between wet (x-axis) and dry (y-axis) habitats. Orange points indicate homeologs with significant shifts detected by HOBIT, while gray points represent homeologs without significant changes.
</p>
</div>
</div>
<div id="homeoroq" class="section level3 hasAnchor" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> HomeoRoq<a href="allohexaploid-c.-flexuosa.html#homeoroq" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For allopolyploids composed of two subgenomes under two different conditions,
HomeoRoq is also a suitable option for detecting homeologs with shifts in expression ratios.
It can be executed using the <code>homeoroq()</code> function in the same way as <code>hobit()</code>,
as shown below.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="allohexaploid-c.-flexuosa.html#cb23-1" tabindex="-1"></a>x_homeoroq <span class="ot">&lt;-</span> <span class="fu">homeoroq</span>(x)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="allohexaploid-c.-flexuosa.html#cb24-1" tabindex="-1"></a><span class="fu">head</span>(x_homeoroq[<span class="fu">order</span>(x_homeoroq<span class="sc">$</span>pvalue), ])</span></code></pre></div>
<pre><code>##              gene pvalue qvalue sumexp__wet__C_hirsuta sumexp__wet__C_amara sumexp__dry__C_hirsuta sumexp__dry__C_amara ratio__wet__C_hirsuta ratio__dry__C_hirsuta  ratio_sd
## 12  CARHR002710_H      0      0               108.3058             43.55477                0.00000             122.9599             0.7131924             0.0000000 0.4620698
## 36  CARHR007170_H      0      0               126.8994             73.38890               12.23196             107.3753             0.6335838             0.1428620 0.2653866
## 52  CARHR011940_H      0      0              1373.8547           1516.85937             1437.42566             394.6598             0.4752648             0.4865562 0.1539877
## 127 CARHR029790_H      0      0               908.8552            775.44869              918.46735            1816.7147             0.5396029             0.5422154 0.1031771
## 346 CARHR099310_H      0      0               458.0346            476.57672              116.10371             558.3269             0.4900803             0.1958960 0.1671791
## 481 CARHR139340_H      0      0               358.4102            992.02619              768.18035             828.8006             0.2654033             0.4364149 0.2223759</code></pre>
<p>The overlap of homeologs with shifts in expression ratios
detected by HOBIT and HomeoRoq can be visualized using an UpSet plot,
which is implemented in the <span class="pkg-name"><a href="https://CRAN.R-project.org/package=UpSetR" target="_blank">UpSetR</a></span> package <span class="citation">(Conway et al. 2017)</span>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="allohexaploid-c.-flexuosa.html#cb26-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;UpSetR&quot;</span>)</span>
<span id="cb26-2"><a href="allohexaploid-c.-flexuosa.html#cb26-2" tabindex="-1"></a>sig_homeologs <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb26-3"><a href="allohexaploid-c.-flexuosa.html#cb26-3" tabindex="-1"></a>    <span class="at">HOBIT =</span> x_output<span class="sc">$</span>gene[x_output<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>],</span>
<span id="cb26-4"><a href="allohexaploid-c.-flexuosa.html#cb26-4" tabindex="-1"></a>    <span class="at">HomeoRoq =</span> x_homeoroq<span class="sc">$</span>gene[x_homeoroq<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>]</span>
<span id="cb26-5"><a href="allohexaploid-c.-flexuosa.html#cb26-5" tabindex="-1"></a>)</span>
<span id="cb26-6"><a href="allohexaploid-c.-flexuosa.html#cb26-6" tabindex="-1"></a></span>
<span id="cb26-7"><a href="allohexaploid-c.-flexuosa.html#cb26-7" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(sig_homeologs))</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-cflex-overlaps"></span>
<img src="_main_files/figure-html/tt-cflex-overlaps-1.png" alt="Overlap of homeologs with shifted expression ratios detected by HOBIT and HomeoRoq." width="672" />
<p class="caption">
Figure 3.3: Overlap of homeologs with shifted expression ratios detected by HOBIT and HomeoRoq.
</p>
</div>
<p>Additionally, homeologs with shifts in expression ratios detected by both methods
can be visualized using the <code>plot_HER()</code> function.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="allohexaploid-c.-flexuosa.html#cb27-1" tabindex="-1"></a>is_sig <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">&quot;n.s.&quot;</span>, <span class="at">length =</span> <span class="fu">nrow</span>(x_output))</span>
<span id="cb27-2"><a href="allohexaploid-c.-flexuosa.html#cb27-2" tabindex="-1"></a>is_sig[x_output<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>] <span class="ot">&lt;-</span> <span class="st">&quot;HOBIT&quot;</span></span>
<span id="cb27-3"><a href="allohexaploid-c.-flexuosa.html#cb27-3" tabindex="-1"></a>is_sig[x_homeoroq<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>] <span class="ot">&lt;-</span> <span class="st">&quot;HomeoRoq&quot;</span></span>
<span id="cb27-4"><a href="allohexaploid-c.-flexuosa.html#cb27-4" tabindex="-1"></a>is_sig[x_output<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">&amp;</span> x_homeoroq<span class="sc">$</span>qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>] <span class="ot">&lt;-</span> <span class="st">&quot;HOBIT&amp;HomeoRoq&quot;</span></span>
<span id="cb27-5"><a href="allohexaploid-c.-flexuosa.html#cb27-5" tabindex="-1"></a>is_sig <span class="ot">&lt;-</span> <span class="fu">factor</span>(is_sig, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;n.s.&quot;</span>, <span class="st">&quot;HOBIT&quot;</span>, <span class="st">&quot;HomeoRoq&quot;</span>, <span class="st">&quot;HOBIT&amp;HomeoRoq&quot;</span>))</span>
<span id="cb27-6"><a href="allohexaploid-c.-flexuosa.html#cb27-6" tabindex="-1"></a><span class="fu">plot_HER</span>(x, <span class="at">label =</span> is_sig)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:tt-cflex-overlaps-highlight"></span>
<img src="_main_files/figure-html/tt-cflex-overlaps-highlight-1.png" alt="Comparison of homeologs with changes in expression ratios detected by HOBIT and HomeoRoq. Each point represents a homeolog expression ratio in wet (x-axis) and dry (y-axis) conditions. Points labeled as n.s. indicate homeologs without significant changes detected by either method; other points represent homeologs identified as having significant changes by the respective method." width="672" />
<p class="caption">
Figure 3.4: Comparison of homeologs with changes in expression ratios detected by HOBIT and HomeoRoq. Each point represents a homeolog expression ratio in wet (x-axis) and dry (y-axis) conditions. Points labeled as n.s. indicate homeologs without significant changes detected by either method; other points represent homeologs identified as having significant changes by the respective method.
</p>
</div>
</div>
<div id="analysis-environment" class="section level3 hasAnchor" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Analysis Environment<a href="allohexaploid-c.-flexuosa.html#analysis-environment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="allohexaploid-c.-flexuosa.html#cb28-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.0 (2025-04-11)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sequoia 15.6.1
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: Asia/Tokyo
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] pROC_1.18.5     UpSetR_1.4.0    gridExtra_2.3   ggplot2_3.5.2   hespresso_1.0.0
## 
## loaded via a namespace (and not attached):
##  [1] tensorA_0.36.2.1     sass_0.4.10          generics_0.1.4       lattice_0.22-7       digest_0.6.37        magrittr_2.0.3       evaluate_1.0.3       grid_4.5.0           RColorBrewer_1.1-3   bookdown_0.44        iterators_1.0.14     fastmap_1.2.0        plyr_1.8.9           foreach_1.5.2        doParallel_1.0.17    jsonlite_2.0.0       processx_3.8.6       backports_1.5.0      limma_3.64.1         ps_1.9.1             scales_1.4.0         codetools_0.2-20     jquerylib_0.1.4      abind_1.4-8          cli_3.6.5            rlang_1.1.6          cmdstanr_0.9.0       splines_4.5.0        withr_3.0.2          cachem_1.1.0         yaml_2.3.10          tools_4.5.0          parallel_4.5.0       checkmate_2.3.2      dplyr_1.1.4          locfit_1.5-9.12      vctrs_0.6.5          posterior_1.6.1      R6_2.6.1             lifecycle_1.0.4      edgeR_4.6.2          pkgconfig_2.0.3      progressr_0.15.1     pillar_1.10.2        bslib_0.9.0          gtable_0.3.6         Rcpp_1.1.0           glue_1.8.0          
## [49] statmod_1.5.0        tidyselect_1.2.1     xfun_0.52            tibble_3.2.1         rstudioapi_0.17.1    knitr_1.50           farver_2.1.2         htmltools_0.5.8.1    labeling_0.4.3       rmarkdown_2.29       compiler_4.5.0       distributional_0.5.0</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="tutorials.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="allohexaploid-t.-aestivum.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  },
  "number_sections": true,
  "biblio-style": "apalike",
  "link-citations": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
