<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.1 HOBIT | hespresso ☕</title>
  <meta name="description" content="4.1 HOBIT | hespresso ☕" />
  <meta name="generator" content="bookdown 0.44 and GitBook 2.6.7" />

  <meta property="og:title" content="4.1 HOBIT | hespresso ☕" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.1 HOBIT | hespresso ☕" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="functions.html"/>
<link rel="next" href="homeoroq-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="_style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>hespresso</a></li>
<li class="chapter" data-level="1" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>1</b> Installation</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a></li>
<li class="chapter" data-level="3" data-path="tutorials.html"><a href="tutorials.html"><i class="fa fa-check"></i><b>3</b> Tutorials</a>
<ul>
<li class="chapter" data-level="3.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html"><i class="fa fa-check"></i><b>3.1</b> Allohexaploid <em>C. flexuosa</em></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#introduction"><i class="fa fa-check"></i><b>3.1.1</b> Introduction</a></li>
<li class="chapter" data-level="3.1.2" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#data-preparation"><i class="fa fa-check"></i><b>3.1.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.1.3" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#hobit"><i class="fa fa-check"></i><b>3.1.3</b> HOBIT</a></li>
<li class="chapter" data-level="3.1.4" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#homeoroq"><i class="fa fa-check"></i><b>3.1.4</b> HomeoRoq</a></li>
<li class="chapter" data-level="3.1.5" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#analysis-environment"><i class="fa fa-check"></i><b>3.1.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html"><i class="fa fa-check"></i><b>3.2</b> Allohexaploid <em>T. aestivum</em></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#introduction-1"><i class="fa fa-check"></i><b>3.2.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#data-preparation-1"><i class="fa fa-check"></i><b>3.2.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.2.3" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#statistical-test"><i class="fa fa-check"></i><b>3.2.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.2.4" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#combining-subgenomes"><i class="fa fa-check"></i><b>3.2.4</b> Combining Subgenomes</a></li>
<li class="chapter" data-level="3.2.5" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#analysis-environment-1"><i class="fa fa-check"></i><b>3.2.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html"><i class="fa fa-check"></i><b>3.3</b> Allotriploid <em>C. insueta</em></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#introduction-2"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#data-preparation-2"><i class="fa fa-check"></i><b>3.3.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#statistical-test-1"><i class="fa fa-check"></i><b>3.3.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.3.4" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#analysis-environment-2"><i class="fa fa-check"></i><b>3.3.4</b> Analysis Environment</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4</b> Functions</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hobit-1.html"><a href="hobit-1.html"><i class="fa fa-check"></i><b>4.1</b> HOBIT</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="hobit-1.html"><a href="hobit-1.html#inputs"><i class="fa fa-check"></i><b>4.1.1</b> Inputs</a></li>
<li class="chapter" data-level="4.1.2" data-path="hobit-1.html"><a href="hobit-1.html#test"><i class="fa fa-check"></i><b>4.1.2</b> Test</a></li>
<li class="chapter" data-level="4.1.3" data-path="hobit-1.html"><a href="hobit-1.html#outputs"><i class="fa fa-check"></i><b>4.1.3</b> Outputs</a></li>
<li class="chapter" data-level="4.1.4" data-path="hobit-1.html"><a href="hobit-1.html#options"><i class="fa fa-check"></i><b>4.1.4</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html"><i class="fa fa-check"></i><b>4.2</b> HomeoRoq</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="homeoroq-1.html"><a href="homeoroq-1.html#inputs-1"><i class="fa fa-check"></i><b>4.2.1</b> Inputs</a></li>
<li class="chapter" data-level="4.2.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html#test-1"><i class="fa fa-check"></i><b>4.2.2</b> Test</a></li>
<li class="chapter" data-level="4.2.3" data-path="homeoroq-1.html"><a href="homeoroq-1.html#outputs-1"><i class="fa fa-check"></i><b>4.2.3</b> Outputs</a></li>
<li class="chapter" data-level="4.2.4" data-path="homeoroq-1.html"><a href="homeoroq-1.html#options-1"><i class="fa fa-check"></i><b>4.2.4</b> Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="evaluation-framework.html"><a href="evaluation-framework.html"><i class="fa fa-check"></i><b>5</b> Evaluation Framework</a>
<ul>
<li class="chapter" data-level="5.1" data-path="simulating-allotetraploid-data.html"><a href="simulating-allotetraploid-data.html"><i class="fa fa-check"></i><b>5.1</b> Simulating Allotetraploid Data</a></li>
<li class="chapter" data-level="5.2" data-path="simulating-allohexaploid-data.html"><a href="simulating-allohexaploid-data.html"><i class="fa fa-check"></i><b>5.2</b> Simulating Allohexaploid Data</a></li>
<li class="chapter" data-level="5.3" data-path="custom-data-simulation.html"><a href="custom-data-simulation.html"><i class="fa fa-check"></i><b>5.3</b> Custom Data Simulation</a></li>
<li class="chapter" data-level="5.4" data-path="evaluation-framework-1.html"><a href="evaluation-framework-1.html"><i class="fa fa-check"></i><b>5.4</b> Evaluation Framework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="citations.html"><a href="citations.html"><i class="fa fa-check"></i>Citations</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><em>hespresso</em> ☕</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="hobit-1" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> HOBIT<a href="hobit-1.html#hobit-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>HOBIT <span class="citation">(Sun et al. 2025)</span> is a statistical method designed
to detect homeologs whose expression ratios differ across multiple conditions.
It employs a likelihood ratio test (LRT) to compare two hierarchical models:
a full model that allows homeolog expression ratios to vary among conditions
and a reduced model that assumes constant ratios across all conditions.</p>
<p>HOBIT models homeolog expression with a negative binomial (NB) distribution.
Specifically, the expression level of a homeolog on the <span class="math inline">\(i\)</span>-th subgenome,
denoted as <span class="math inline">\(x_i\)</span>, is assumed to follow an NB distribution
with mean <span class="math inline">\(\mu\theta_i\)</span> and dispersion <span class="math inline">\(\phi\)</span>,</p>
<p><span class="math display">\[
x_{i} \sim NB(\mu\theta_{i}, \phi)
\]</span></p>
<p>where <span class="math inline">\(\mu\)</span> represents the total gene expression across all subgenomes
and <span class="math inline">\(\theta_{i}\)</span> denotes the homeolog expression ratio for the <span class="math inline">\(i\)</span>-th subgenome.</p>
<p>By default, HOBIT samples the parameters
<span class="math inline">\(\mu\)</span> and <span class="math inline">\(\boldsymbol{\theta} = \left(\theta_{1}, \theta_{2}, \cdots \right)\)</span>
from non-informative prior (NIP) distributions.
The dispersion parameter <span class="math inline">\(\phi\)</span> is estimated using the <code>estimateDisp()</code> function
from the <span class="pkg-name"><a href="https://bioconductor.org/packages/edgeR" target="_blank">edgeR</a></span> package <span class="citation">(Chen et al. 2025)</span>.
Based on these assumptions,
HOBIT performs Markov chain Monte Carlo (MCMC) sampling
to fit both the full and reduced models
and to compute their respective likelihoods.
Then, a LRT is conducted using the likelihoods derived from the two models.</p>
<p>The <code>hobit()</code> function provides an end-to-end interface for parameter estimation,
model fitting, likelihood computation, and hypothesis testing.
It internally uses the <code>sample()</code> function from the <span class="pkg-name"><a href="https://mc-stan.org/cmdstanr" target="_blank">cmdstanr</a></span> package <span class="citation">(Gabry et al. 2025)</span>
to perform MCMC sampling,
inheriting its configuration options for flexible and scalable inference.</p>
<div id="inputs" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Inputs<a href="hobit-1.html#inputs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <code>hobit()</code> function requires an input object of class <code>ExpMX</code>.
To create this object from real RNA-Seq data,
users need to provide the following three components:</p>
<ul>
<li>a gene expression matrix</li>
<li>a vector or data.frame of condition labels</li>
<li>a mapping table to organize gene expression data into homeolog expression</li>
</ul>
<p>The following example shows how to load gene expression data,
define experimental conditions, and load the mapping table from a file.</p>
<p>The gene expression matrix should be a data.frame or matrix,
with row names corresponding to gene names.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="hobit-1.html#cb70-1" tabindex="-1"></a>gexp <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.0516.mini.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb70-2"><a href="hobit-1.html#cb70-2" tabindex="-1"></a><span class="fu">head</span>(gexp)</span></code></pre></div>
<pre><code>##               wet_1 wet_2 wet_3 dry_1 dry_2 dry_3
## CARHR000350_H    21     8    17    19    10    14
## CARHR000580_H    24    21    14    26    11    19
## CARHR001360_H   440   264   364   394   272   248
## CARHR001380_H    36   109    48    52    52    32
## CARHR001400_H    10    11    59    30    23    27
## CARHR001480_H    27    17    49    40    25    15</code></pre>
<p>The condition labels can be provided as a vector or a data.frame.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="hobit-1.html#cb72-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>)</span>
<span id="cb72-2"><a href="hobit-1.html#cb72-2" tabindex="-1"></a>group</span></code></pre></div>
<pre><code>## [1] &quot;wet&quot; &quot;wet&quot; &quot;wet&quot; &quot;dry&quot; &quot;dry&quot; &quot;dry&quot;</code></pre>
<p>The mapping table (<code>mapping_table</code>) is a data.frame
in which the first and second columns correspond to gene names
from <em>C. hirsuta</em> and <em>C. amara</em>, respectively.
It is assumed that all gene names in both columns appear
in the row names of gene expression matrix (<code>gexp</code>).</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="hobit-1.html#cb74-1" tabindex="-1"></a>mapping_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.homeolog.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb74-2"><a href="hobit-1.html#cb74-2" tabindex="-1"></a><span class="fu">head</span>(mapping_table)</span></code></pre></div>
<pre><code>##       C_hirsuta       C_amara
## 1 CARHR000010_H CARHR000010_A
## 2 CARHR000060_H CARHR000060_A
## 3 CARHR000070_H CARHR000070_A
## 4 CARHR000090_H CARHR000090_A
## 5 CARHR000110_H CARHR000110_A
## 6 CARHR000120_H CARHR000120_A</code></pre>
<p>Once the three required components are prepared,
an <code>ExpMX</code> object can be created using the <code>newExpMX()</code> function.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="hobit-1.html#cb76-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">newExpMX</span>(gexp, group, mapping_table)</span>
<span id="cb76-2"><a href="hobit-1.html#cb76-2" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>## # 2 subgenomes (C_hirsuta, C_amara)
## # 1000 homeologs
## ---------------------
## Experiment Design:
##   group
## 1   wet
## 2   wet
## 3   wet
## 4   dry
## 5   dry
## 6   dry
## ---------------------
## &gt; subgenome: C_hirsuta 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]   21    8   17   19   10   14
## [2,]   24   21   14   26   11   19
## [3,]  440  264  364  394  272  248
## [4,]   36  109   48   52   52   32
## [5,]   10   11   59   30   23   27
## [6,]   27   17   49   40   25   15
## +++++++++++++++++++++
## &gt; subgenome: C_amara 
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]  152  168  183  151   92   75
## [2,]   54   44   49   46   22   27
## [3,]  344  181  242  257  203  183
## [4,]   73  150   49   43   37   32
## [5,]    4    0    2    2    6    1
## [6,]   30   23  124   48   32   28
## ---------------------</code></pre>
</div>
<div id="test" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Test<a href="hobit-1.html#test" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Usually, normalization is required to correct for differences in library sizes
across replicates.
If the given expression data (<code>gexp</code>) have not been normalized,
apply an appropriate normalization method before analysis.
In this example, we normalize the data using the <code>norm_counts()</code> function
provided in the <span class="pkg-name"><a href="https://bitdessin.github.io/hespresso" target="_blank">hespresso</a></span> package.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="hobit-1.html#cb78-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">norm_counts</span>(x)</span></code></pre></div>
<p>By default, <code>norm_counts()</code> function performs TMM normalization
by internally calling the <span class="pkg-name"><a href="https://bioconductor.org/packages/edgeR" target="_blank">edgeR</a></span> package <span class="citation">(Robinson and Oshlack 2010; Chen et al. 2025)</span>.
However, users may choose to perform normalization using other appropriate
packages such as <span class="pkg-name"><a href="https://bioconductor.org/packages/DESeq2" target="_blank">DESeq2</a></span> <span class="citation">(Love et al. 2014)</span>.
Note that, this step can be skipped
if the expression data has already been normalized, such as FPKM and TPM.</p>
<p>Once the <code>ExpMX</code> object has been prepared and the data have been normalized,
homeologs with shifted expression ratios
can be detected using HOBIT by running the <code>hobit()</code> function, as shown below.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="hobit-1.html#cb79-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x)</span></code></pre></div>
</div>
<div id="outputs" class="section level3 hasAnchor" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Outputs<a href="hobit-1.html#outputs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The result (<code>x_output</code>) is a <code>data.frame</code> object, with one row per homeolog.
The row order corresponds to the input,
and the columns summarize the results of the statistical test.</p>
<ul>
<li><code>gene</code>: Homeolog ID. The gene names defined in the first column of the mapping table.</li>
<li><code>pvalue</code>: <em>p</em>-value from the LRT using normalized likelihoods.</li>
<li><code>qvalue</code>: Adjusted <em>p</em>-value using the Benjamini-Hochberg method.</li>
<li><code>raw_pvalue</code>: <em>p</em>-value from the LRT using raw likelihoods.</li>
<li><code>raw_qvalue</code>: Adjusted raw <em>p</em>-value using the Benjamini-Hochberg method.</li>
<li><code>Dmax</code>: Maximum absolute difference of homeolog expression ratios between conditions.</li>
<li><code>ORmax</code>: Maximum odds ratio of homeolog expression ratios between conditions.</li>
<li><code>theta0__&lt;s&gt;</code>: Posterior homeolog expression ratio estimates shared across all conditions, where <code>&lt;s&gt;</code> denotes the subgenome name. These parameters are used to compute the log-likelihood of the reduced model.</li>
<li><code>theta__&lt;s&gt;__&lt;g&gt;</code>: Posterior homeolog expression ratio estimates specific to each condition, where <code>&lt;s&gt;</code> denotes the subgenome name and and <code>&lt;g&gt;</code> denotes the group name. These parameters are used to compute the log-likelihood of the full model.</li>
<li><code>logLik_H1</code>: Log-likelihood of the full model.</li>
<li><code>logLik_H0</code>: Log-likelihood of the reduced model.</li>
</ul>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="hobit-1.html#cb80-1" tabindex="-1"></a><span class="fu">head</span>(x_output)</span></code></pre></div>
<pre><code>##            gene    pvalue qvalue raw_pvalue raw_qvalue       Dmax    ORmax theta__C_hirsuta__dry theta__C_amara__dry theta__C_hirsuta__wet theta__C_amara__wet theta0__C_hirsuta theta0__C_amara logLik_H1 logLik_H0
## 1 CARHR000350_H 0.3329165      1  0.2469988          1 0.04285630 1.632985             0.1237820           0.8762180            0.07984355          0.92015650         0.1032560       0.8967440 -47.12410 -47.77205
## 2 CARHR000580_H 0.6029891      1  0.5340392          1 0.08695550 1.483685             0.3778220           0.6221780            0.29002300          0.70997700         0.3377145       0.6622855 -42.34160 -42.42485
## 3 CARHR001360_H 1.0000000      1  1.0000000          1 0.00078200 1.003275             0.5802315           0.4197685            0.58026950          0.41973050         0.5801480       0.4198520 -67.37725 -67.02965
## 4 CARHR001380_H 0.3502746      1  0.2640882          1 0.13138900 1.713870             0.5479325           0.4520675            0.41441100          0.58558900         0.4822940       0.5177060 -68.00180 -68.77350
## 5 CARHR001400_H 0.6444191      1  0.5810750          1 0.04599315 1.812507             0.8838030           0.1161970            0.93227600          0.06772415         0.9037260       0.0962739 -41.85440 -41.94350
## 6 CARHR001480_H 0.9067870      1  0.8886615          1 0.05771650 1.279647             0.4142085           0.5857915            0.35662800          0.64337200         0.3841715       0.6158285 -54.01690 -53.90250</code></pre>
</div>
<div id="options" class="section level3 hasAnchor" number="4.1.4">
<h3><span class="header-section-number">4.1.4</span> Options<a href="hobit-1.html#options" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="basic-options" class="section level4 hasAnchor" number="4.1.4.1">
<h4><span class="header-section-number">4.1.4.1</span> Basic Options<a href="hobit-1.html#basic-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Below is a basic example of running the <code>hobit()</code> function.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="hobit-1.html#cb82-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x)</span></code></pre></div>
<p>By default, <code>hobit()</code> uses multiple threads,
defined by the <code>mc.cores</code> option in the global environment, for MCMC sampling,
If <code>mc.cores</code> is not set, it defaults to using a single thread.
Users can customize the number of threads to accelerate the testing process
by either setting <code>mc.cores</code> (e.g., <code>options(mc.cores = 8)</code>)
or specifying the <code>n_threads</code> argument directly. For example:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="hobit-1.html#cb83-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">n_threads =</span> <span class="dv">8</span>)</span></code></pre></div>
<p>Additionally, user can control parallelization of the MCMC sampling process via the <code>parallel_chains</code> argument.
When specified, this argument is passed directly to the <code>sample()</code> function from the
<span class="pkg-name"><a href="https://mc-stan.org/cmdstanr" target="_blank">cmdstanr</a></span> package, allowing parallel execution of MCMC chains.</p>
<p>For example, to run four parallel chains, each on a separate thread:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="hobit-1.html#cb84-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">n_threads =</span> <span class="dv">1</span>, <span class="at">parallel_chains =</span> <span class="dv">4</span>, <span class="at">chains =</span> <span class="dv">4</span>)</span></code></pre></div>
<p>Note that the <code>n_threads</code> and <code>parallel_chains</code> arguments
control parallelization at different stages.
The <code>n_threads</code> argument controls parallelization during the processing of homeologs,
while <code>parallel_chains</code> controls parallelization during MCMC chains.
Hence, if both are specified, users should reserve <code>n_threads</code> × <code>parallel_chains</code> threads.</p>
</div>
<div id="model-options" class="section level4 hasAnchor" number="4.1.4.2">
<h4><span class="header-section-number">4.1.4.2</span> Model Options<a href="hobit-1.html#model-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>By default, HOBIT assumes that homeolog expression follows a NB distribution.
In this model, the mean is sampled from a NIP distribution,
and the dispersion parameter is estimated using the <code>estimateDisp()</code> function
from the <span class="pkg-name"><a href="https://bioconductor.org/packages/edgeR" target="_blank">edgeR</a></span> package <span class="citation">(Chen et al. 2025)</span>.
In addition to these defaults,
HOBIT provides options for users to customize these assumptions as needed.</p>
<p>To switch from the default NB distribution to a zero-inflated negative binomial
(ZINB) distribution, user can set <code>dist = "ZINB"</code>.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="hobit-1.html#cb85-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">dist =</span> <span class="st">&quot;ZINB&quot;</span>)</span></code></pre></div>
<p>By default, homeolog expression ratios are sampled from NIP distributions,
with the constraint that their sum equals 1.
Beside the defaults, HOBIT also supports using a Dirichlet prior distribution,
with parameters estimated from the observed homeolog expression values.
This can be enabled by setting the <code>use_Dirichlet</code> option, as shown below.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="hobit-1.html#cb86-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">use_Dirichlet =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The default dispersion estimation method requires multiple biological replicates per group.
If the dataset contains only a single replicate per group, dispersion estimation will fail.
In such cases, HOBIT provides the <code>no_replicate</code> option.
When set to <code>TRUE</code>, all replicates are treated as if they belong to a single group for dispersion estimation.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="hobit-1.html#cb87-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">no_replicate =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Note that while dispersion estimation without replicates can be enabled by setting <code>no_replicate = TRUE</code>,
this approach is not recommended for biological experiments.
For reliable statistical inference,
we strongly encourage reconsidering the experimental design to include multiple biological replicates per group.</p>
</div>
<div id="mcmc-options" class="section level4 hasAnchor" number="4.1.4.3">
<h4><span class="header-section-number">4.1.4.3</span> MCMC Options<a href="hobit-1.html#mcmc-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Several options control the behavior of MCMC sampling and
can significantly affect the accuracy and efficiency of inference.
Although <code>hobit()</code> supports all options available in the <code>sample()</code> function
from the <span class="pkg-name"><a href="https://mc-stan.org/cmdstanr" target="_blank">cmdstanr</a></span> package,
the following are particularly important.</p>
<ul>
<li><code>chains</code>: Number of MCMC chains to run (default: <code>4</code>).</li>
<li><code>iter_warmup</code>: Number of warmup iterations per chain (default: <code>1000</code>).</li>
<li><code>iter_sampling</code>: Number of post-warmup iterations per chain (default: <code>1000</code>).</li>
<li><code>save_warmup</code>: Whether to save warmup iterations (default: <code>FALSE</code>).</li>
<li><code>thin</code>: Interval at which samples are saved (default: <code>1</code>). Typically unchanged unless memory issues arise.</li>
<li><code>max_treedepth</code>: Maximum tree depth for the NUTS sampler (default: <code>10</code>). Increasing it may help resolve convergence issues but slows execution.</li>
<li><code>adapt_delta</code>: Target acceptance rate for the sampler (default: <code>0.8</code>). Higher values improve stability but may reduce speed.</li>
</ul>
<p>The following example increases the number of chains and iterations to improve stability and convergence,
though at the cost of longer execution times:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="hobit-1.html#cb88-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">chains =</span> <span class="dv">8</span>, <span class="at">iter_warmup =</span> <span class="dv">2000</span>, <span class="at">iter_sampling =</span> <span class="dv">10000</span>)</span></code></pre></div>
<p>If warnings or convergence issues arise with the default settings,
consider increasing <code>max_treedepth</code> and <code>adapt_delta</code>.
Note that increasing these parameters may result in slower sampling.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="hobit-1.html#cb89-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">max_treedepth =</span> <span class="dv">12</span>, <span class="at">adapt_delta =</span> <span class="fl">0.95</span>)</span></code></pre></div>
<p>For more advanced configuration options, refer to the
<a href="https://mc-stan.org/cmdstanr/reference/model-method-sample.html">cmdstanr documentation</a>.</p>
</div>
<div id="verbosity-options" class="section level4 hasAnchor" number="4.1.4.4">
<h4><span class="header-section-number">4.1.4.4</span> Verbosity Options<a href="hobit-1.html#verbosity-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Verbosity options are available to control the verbosity of progress messages during sampling.</p>
<ul>
<li><code>refresh</code>: Frequency of progress updates to the console (default: <code>100</code>). Set to <code>0</code> to suppress updates.</li>
<li><code>show_messages</code>: Whether to print all diagnostic and progress messages (default: <code>TRUE</code>).</li>
</ul>
<p>To reduce console output during execution, adjust these settings as shown below.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="hobit-1.html#cb90-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">hobit</span>(x, <span class="at">refresh =</span> <span class="dv">0</span>, <span class="at">show_messages =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="functions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="homeoroq-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  },
  "number_sections": true,
  "biblio-style": "apalike",
  "link-citations": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
