<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.2 HomeoRoq | hespresso ☕</title>
  <meta name="description" content="4.2 HomeoRoq | hespresso ☕" />
  <meta name="generator" content="bookdown 0.44 and GitBook 2.6.7" />

  <meta property="og:title" content="4.2 HomeoRoq | hespresso ☕" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.2 HomeoRoq | hespresso ☕" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="hobit-1.html"/>
<link rel="next" href="evaluation-framework.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="_style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>hespresso</a></li>
<li class="chapter" data-level="1" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>1</b> Installation</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a></li>
<li class="chapter" data-level="3" data-path="tutorials.html"><a href="tutorials.html"><i class="fa fa-check"></i><b>3</b> Tutorials</a>
<ul>
<li class="chapter" data-level="3.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html"><i class="fa fa-check"></i><b>3.1</b> Allohexaploid <em>C. flexuosa</em></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#introduction"><i class="fa fa-check"></i><b>3.1.1</b> Introduction</a></li>
<li class="chapter" data-level="3.1.2" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#data-preparation"><i class="fa fa-check"></i><b>3.1.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.1.3" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#hobit"><i class="fa fa-check"></i><b>3.1.3</b> HOBIT</a></li>
<li class="chapter" data-level="3.1.4" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#homeoroq"><i class="fa fa-check"></i><b>3.1.4</b> HomeoRoq</a></li>
<li class="chapter" data-level="3.1.5" data-path="allohexaploid-c.-flexuosa.html"><a href="allohexaploid-c.-flexuosa.html#analysis-environment"><i class="fa fa-check"></i><b>3.1.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html"><i class="fa fa-check"></i><b>3.2</b> Allohexaploid <em>T. aestivum</em></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#introduction-1"><i class="fa fa-check"></i><b>3.2.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2.2" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#data-preparation-1"><i class="fa fa-check"></i><b>3.2.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.2.3" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#statistical-test"><i class="fa fa-check"></i><b>3.2.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.2.4" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#combining-subgenomes"><i class="fa fa-check"></i><b>3.2.4</b> Combining Subgenomes</a></li>
<li class="chapter" data-level="3.2.5" data-path="allohexaploid-t.-aestivum.html"><a href="allohexaploid-t.-aestivum.html#analysis-environment-1"><i class="fa fa-check"></i><b>3.2.5</b> Analysis Environment</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html"><i class="fa fa-check"></i><b>3.3</b> Allotriploid <em>C. insueta</em></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#introduction-2"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#data-preparation-2"><i class="fa fa-check"></i><b>3.3.2</b> Data Preparation</a></li>
<li class="chapter" data-level="3.3.3" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#statistical-test-1"><i class="fa fa-check"></i><b>3.3.3</b> Statistical Test</a></li>
<li class="chapter" data-level="3.3.4" data-path="allotriploid-c.-insueta.html"><a href="allotriploid-c.-insueta.html#analysis-environment-2"><i class="fa fa-check"></i><b>3.3.4</b> Analysis Environment</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4</b> Functions</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hobit-1.html"><a href="hobit-1.html"><i class="fa fa-check"></i><b>4.1</b> HOBIT</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="hobit-1.html"><a href="hobit-1.html#inputs"><i class="fa fa-check"></i><b>4.1.1</b> Inputs</a></li>
<li class="chapter" data-level="4.1.2" data-path="hobit-1.html"><a href="hobit-1.html#test"><i class="fa fa-check"></i><b>4.1.2</b> Test</a></li>
<li class="chapter" data-level="4.1.3" data-path="hobit-1.html"><a href="hobit-1.html#outputs"><i class="fa fa-check"></i><b>4.1.3</b> Outputs</a></li>
<li class="chapter" data-level="4.1.4" data-path="hobit-1.html"><a href="hobit-1.html#options"><i class="fa fa-check"></i><b>4.1.4</b> Options</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html"><i class="fa fa-check"></i><b>4.2</b> HomeoRoq</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="homeoroq-1.html"><a href="homeoroq-1.html#inputs-1"><i class="fa fa-check"></i><b>4.2.1</b> Inputs</a></li>
<li class="chapter" data-level="4.2.2" data-path="homeoroq-1.html"><a href="homeoroq-1.html#test-1"><i class="fa fa-check"></i><b>4.2.2</b> Test</a></li>
<li class="chapter" data-level="4.2.3" data-path="homeoroq-1.html"><a href="homeoroq-1.html#outputs-1"><i class="fa fa-check"></i><b>4.2.3</b> Outputs</a></li>
<li class="chapter" data-level="4.2.4" data-path="homeoroq-1.html"><a href="homeoroq-1.html#options-1"><i class="fa fa-check"></i><b>4.2.4</b> Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="evaluation-framework.html"><a href="evaluation-framework.html"><i class="fa fa-check"></i><b>5</b> Evaluation Framework</a>
<ul>
<li class="chapter" data-level="5.1" data-path="simulating-allotetraploid-data.html"><a href="simulating-allotetraploid-data.html"><i class="fa fa-check"></i><b>5.1</b> Simulating Allotetraploid Data</a></li>
<li class="chapter" data-level="5.2" data-path="simulating-allohexaploid-data.html"><a href="simulating-allohexaploid-data.html"><i class="fa fa-check"></i><b>5.2</b> Simulating Allohexaploid Data</a></li>
<li class="chapter" data-level="5.3" data-path="custom-data-simulation.html"><a href="custom-data-simulation.html"><i class="fa fa-check"></i><b>5.3</b> Custom Data Simulation</a></li>
<li class="chapter" data-level="5.4" data-path="evaluation-framework-1.html"><a href="evaluation-framework-1.html"><i class="fa fa-check"></i><b>5.4</b> Evaluation Framework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="citations.html"><a href="citations.html"><i class="fa fa-check"></i>Citations</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><em>hespresso</em> ☕</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="homeoroq-1" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> HomeoRoq<a href="homeoroq-1.html#homeoroq-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>HomeoRoq <span class="citation">(Akama et al. 2014)</span> estimates the probability that
homeolog expression ratios remain constant
between two experimental groups using RNA-Seq read count data for each homeolog.
Since this probability cannot be derived analytically,
HomeoRoq adopts a Bayesian framework
to approximate the posterior distribution of homeolog expression ratios
under the null hypothesis of no change between conditions.
The <code>homeoroq()</code> function provides a complete interface
for detecting homeologs whose expression ratios differ between two conditions.</p>
<p>By default, it performs 10,000 sampling iterations (<code>iter_sampling = 1e4</code>) per probability calculation
and repeats the calculation four times (<code>chians = 4</code>).
However, for robust and reproducible estimation of statistical significance (i.e., <em>p</em>-values),
it is recommended to run multiple independent sampling replicates,
each with a sufficiently large number of iterations.
To reproduce the settings used in the original HomeoRoq implementation <span class="citation">(Akama et al. 2014)</span>,
set <code>iter_sampling = 1e4</code> and <code>chains = 10</code>.</p>
<div id="inputs-1" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Inputs<a href="homeoroq-1.html#inputs-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <code>homeoroq()</code> function requires an input object of class <code>ExpMX</code>,
same to the <code>hobit()</code> function.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="homeoroq-1.html#cb91-1" tabindex="-1"></a>gexp <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.0516.mini.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb91-2"><a href="homeoroq-1.html#cb91-2" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;wet&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>, <span class="st">&quot;dry&quot;</span>)</span>
<span id="cb91-3"><a href="homeoroq-1.html#cb91-3" tabindex="-1"></a>mapping_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;../data/c_flexuosa.homeolog.txt.gz&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb91-4"><a href="homeoroq-1.html#cb91-4" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">newExpMX</span>(gexp, group, mapping_table)</span></code></pre></div>
</div>
<div id="test-1" class="section level3 hasAnchor" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Test<a href="homeoroq-1.html#test-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Once an <code>ExpMX</code> object is prepared,
normalize the counts if the data have not already been normalized,
then run the test using the <code>homeoroq()</code> function.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="homeoroq-1.html#cb92-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">norm_counts</span>(x)</span>
<span id="cb92-2"><a href="homeoroq-1.html#cb92-2" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">homeoroq</span>(x)</span></code></pre></div>
</div>
<div id="outputs-1" class="section level3 hasAnchor" number="4.2.3">
<h3><span class="header-section-number">4.2.3</span> Outputs<a href="homeoroq-1.html#outputs-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The result (<code>x_output</code>) is a <code>data.frame</code> class object with one row per homeolog,
summarizing the statistical test results:</p>
<ul>
<li><code>gene</code>: Homeolog ID. The gene names defined in the first column of the mapping table.</li>
<li><code>pvalue</code>: <em>p</em>-value from the likelihood ratio test.</li>
<li><code>qvalue</code>: Adjusted <em>p</em>-value using the Benjamini-Hochberg method.</li>
<li><code>sumexp__&lt;s&gt;__&lt;g&gt;</code>: Total read counts for subgenome <code>&lt;s&gt;</code> under condition <code>&lt;g&gt;</code>.</li>
<li><code>ratio__&lt;s&gt;__&lt;g&gt;</code>: HERs for subgenome <code>&lt;s&gt;</code> under condition <code>&lt;g&gt;</code>.</li>
<li><code>ratio_sd</code>: Standard deviation of HERs calculated from observed counts.</li>
</ul>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="homeoroq-1.html#cb93-1" tabindex="-1"></a><span class="fu">head</span>(x_output)</span></code></pre></div>
<pre><code>##            gene   pvalue qvalue sumexp__wet__C_hirsuta sumexp__wet__C_amara sumexp__dry__C_hirsuta sumexp__dry__C_amara ratio__wet__C_hirsuta ratio__dry__C_hirsuta   ratio_sd
## 1 CARHR000350_H 0.564425      1               44.41277           504.994715               43.73729            319.95367            0.08083758            0.07970611 0.03421429
## 2 CARHR000580_H 0.720325      1               60.75627           147.119479               56.29544             94.87458            0.29227204            0.27675177 0.05769653
## 3 CARHR001360_H 0.999925      1             1054.50900           758.465421              937.45041            668.95348            0.58164582            0.55276942 0.01598056
## 4 CARHR001380_H 0.429150      1              207.83502           295.104109              143.04677            117.38436            0.41324091            0.32647834 0.08361259
## 5 CARHR001400_H 0.957500      1               71.75076             5.583816               83.95502             10.24313            0.92779663            0.93763805 0.10320192
## 6 CARHR001480_H 0.996750      1               87.59598           159.502354               79.72566            110.13494            0.35449846            0.33326221 0.06667744</code></pre>
</div>
<div id="options-1" class="section level3 hasAnchor" number="4.2.4">
<h3><span class="header-section-number">4.2.4</span> Options<a href="homeoroq-1.html#options-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The number of iterations and number of samples drawn per iteration can be
adjusted using the <code>n_iters</code> and <code>iter_sampling</code> arguments, respectively.
For example:</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="homeoroq-1.html#cb95-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">homeoroq</span>(x, <span class="at">n_iters =</span> <span class="dv">8</span>, <span class="at">iter_sampling =</span> <span class="fl">1e5</span>)</span></code></pre></div>
<p>Additionally, the <code>n_threads</code> option can be used to run iterations in parallel.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="homeoroq-1.html#cb96-1" tabindex="-1"></a>x_output <span class="ot">&lt;-</span> <span class="fu">homeoroq</span>(x, <span class="at">n_threads =</span> <span class="dv">8</span>, <span class="at">n_iters =</span> <span class="dv">8</span>, <span class="at">iter_sampling =</span> <span class="fl">1e5</span>)</span></code></pre></div>

</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="hobit-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="evaluation-framework.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  },
  "number_sections": true,
  "biblio-style": "apalike",
  "link-citations": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
