## HOBIT

```{r fc-hobit-setup, include=FALSE}
source("../_config.R")
```

HOBIT [@ref_hobit] is a statistical method designed
to detect homeologs whose expression ratios differ across multiple conditions.
It employs a likelihood ratio test (LRT) to compare two hierarchical models:
a full model that allows homeolog expression ratios to vary among conditions
and a reduced model that assumes constant ratios across all conditions.

HOBIT models homeolog expression with a negative binomial (NB) distribution.
Specifically, the expression level of a homeolog on the $i$-th subgenome,
denoted as $x_i$, is assumed to follow an NB distribution
with mean $\mu\theta_i$ and dispersion $\phi$,


$$
x_{i} \sim NB(\mu\theta_{i}, \phi)
$$

where $\mu$ represents the total gene expression across all subgenomes
and $\theta_{i}$ denotes the homeolog expression ratio for the $i$-th subgenome.

By default, HOBIT samples the parameters
$\mu$ and $\boldsymbol{\theta} = \left(\theta_{1}, \theta_{2}, \cdots \right)$
from non-informative prior (NIP) distributions.
The dispersion parameter $\phi$ is estimated using the `estimateDisp()` function
from the `r pkg("edgeR", "bioc")` package [@ref_edger].
Based on these assumptions,
HOBIT performs Markov chain Monte Carlo (MCMC) sampling
to fit both the full and reduced models
and to compute their respective likelihoods.
Then, a LRT is conducted using the likelihoods derived from the two models.

The `hobit()` function provides an end-to-end interface for parameter estimation,
model fitting, likelihood computation, and hypothesis testing.
It internally uses the `sample()` function from the `r pkg("cmdstanr")` package [@ref_stan]
to perform MCMC sampling,
inheriting its configuration options for flexible and scalable inference.


### Inputs

The `hobit()` function requires an input object of class `ExpMX`.
To create this object from real RNA-Seq data,
users need to provide the following three components:

- a gene expression matrix
- a vector or data.frame of condition labels
- a mapping table to organize gene expression data into homeolog expression

The following example shows how to load gene expression data,
define experimental conditions, and load the mapping table from a file.

The gene expression matrix should be a data.frame or matrix,
with row names corresponding to gene names.

```{r fc-hobit-input-preparation-gexp}
gexp <- read.table("../data/c_flexuosa.0516.mini.txt.gz", header = TRUE, sep = "\t", row.names = 1)
head(gexp)
```

The condition labels can be provided as a vector or a data.frame.

```{r fc-hobit-input-preparation-mappingtable}
group <- c("wet", "wet", "wet", "dry", "dry", "dry")
group
```

The mapping table (`mapping_table`) is a data.frame
in which the first and second columns correspond to gene names
from _C. hirsuta_ and _C. amara_, respectively.
It is assumed that all gene names in both columns appear
in the row names of gene expression matrix (`gexp`).

```{r fc-hobit-input-preparation-group}
mapping_table <- read.table("../data/c_flexuosa.homeolog.txt.gz", header = TRUE, sep = "\t")
head(mapping_table)
```

Once the three required components are prepared,
an `ExpMX` object can be created using the `newExpMX()` function.

```{r, fc-hobit-input-newExpMX}
x <- newExpMX(gexp, group, mapping_table)
x
```


### Test

Usually, normalization is required to correct for differences in library sizes
across replicates.
If the given expression data (`gexp`) have not been normalized,
apply an appropriate normalization method before analysis.
In this example, we normalize the data using the `norm_counts()` function
provided in the `r pkg("hespresso")` package.

```{r, fc-hobit-run-norm}
x <- norm_counts(x)
```

By default, `norm_counts()` function performs TMM normalization
by internally calling the `r pkg("edgeR", "bioc")` package [@ref_tmm; @ref_edger].
However, users may choose to perform normalization using other appropriate
packages such as `r pkg("DESeq2", "bioc")` [@ref_deseq2].
Note that, this step can be skipped
if the expression data has already been normalized, such as FPKM and TPM.

Once the `ExpMX` object has been prepared and the data have been normalized,
homeologs with shifted expression ratios
can be detected using HOBIT by running the `hobit()` function, as shown below.

```{r fc-hobit-run-hobit, results="hide", message=FALSE, warning=FALSE}
x_output <- hobit(x)
```


### Outputs

The result (`x_output`) is a `data.frame` with one row per homeolog.
The row order matches the input,
and the columns summarize the results of the statistical test.
Note that all statistics are derived from MCMC samples
and may differ from those calculated directly from the RNA-Seq read counts.

- `gene`: Homeolog ID. The gene names defined in the first column of the mapping table.
- `pvalue`: p-value from the LRT using normalized likelihoods.
- `qvalue`: Benjamini–Hochberg adjusted `pvalue`.
- `raw_pvalue`: p-value from the LRT using raw likelihoods.
- `raw_qvalue`: Benjamini–Hochberg adjusted `raw_pvalue`.
- `D__$__*`: Difference in expression ratios for subgenome `$` between the two groups represented by `*`.
- `OR__$__*`: Odds ratio of expression ratios for subgenome `$` between the two groups represented by `*`.
- `Dmax`: Maximum absolute difference in expression ratios (`D__$__*`) observed across all subgenomes and group comparisons.
- `ORmax`: Maximum odds ratio in expression ratios (`OR__$__*`) observed across all subgenomes and group comparisons.
- `theta0__$`: Posterior expression ratio estimates shared across all conditions, where `$` denotes the subgenome name.
- `theta1__*__$`: Posterior expression ratio estimates specific to each condition, where `$` denotes the subgenome name and `*` denotes the group name.
- `logLik_H0`: Log-likelihood of the reduced model.
- `logLik_H1`: Log-likelihood of the full model.


```{r, fc-hobit-output-print}
head(x_output)
```


### Options

#### Basic Options

Below is a basic example of running the `hobit()` function.

```{r fc-hobit-basicopts-default, eval=FALSE}
x_output <- hobit(x)
```

By default, `hobit()` uses multiple threads,
defined by the `mc.cores` option in the global environment, for MCMC sampling,
If `mc.cores` is not set, it defaults to using a single thread.
Users can customize the number of threads to accelerate the testing process
by either setting `mc.cores` (e.g., `options(mc.cores = 8)`)
or specifying the `n_threads` argument directly. For example:

```{r fc-hobit-basicopts-nthreads, eval=FALSE}
x_output <- hobit(x, n_threads = 8)
```

Additionally, user can control parallelization of the MCMC sampling process via the `parallel_chains` argument.
When specified, this argument is passed directly to the `sample()` function from the
`r pkg("cmdstanr")` package, allowing parallel execution of MCMC chains.

For example, to run four parallel chains, each on a separate thread:

```{r fc-hobit-basicopts-run-parallel, eval=FALSE}
x_output <- hobit(x, n_threads = 1, parallel_chains = 4, chains = 4)
```

Note that the `n_threads` and `parallel_chains` arguments
control parallelization at different stages.
The `n_threads` argument controls parallelization during the processing of homeologs,
while `parallel_chains` controls parallelization during MCMC chains.
Hence, if both are specified, users should reserve `n_threads` × `parallel_chains` threads.


#### Model Options

By default, HOBIT assumes that homeolog expression follows a NB distribution.
In this model, the mean is sampled from a NIP distribution,
and the dispersion parameter is estimated using the `estimateDisp()` function
from the `r pkg("edgeR", "bioc")` package [@ref_edger].
In addition to these defaults,
HOBIT provides options for users to customize these assumptions as needed.

To switch from the default NB distribution to a zero-inflated negative binomial
(ZINB) distribution, user can set `dist = "ZINB"`.

```{r fc-hobit-modelopts-ZINB, eval=FALSE}
x_output <- hobit(x, dist = "ZINB")
```

By default, homeolog expression ratios are sampled from NIP distributions,
with the constraint that their sum equals 1.
Beside the defaults, HOBIT also supports using a Dirichlet prior distribution,
with parameters estimated from the observed homeolog expression values.
This can be enabled by setting the `use_Dirichlet` option, as shown below.

```{r fc-hobit-modelopts-useprior, eval=FALSE}
x_output <- hobit(x, use_Dirichlet = TRUE)
```

The default dispersion estimation method requires multiple biological replicates per group.
If the dataset contains only a single replicate per group, dispersion estimation will fail.
In such cases, HOBIT provides the `no_replicate` option.
When set to `TRUE`, all replicates are treated as if they belong to a single group for dispersion estimation.

```{r fc-hobit-modelopts-norep, eval=FALSE}
x_output <- hobit(x, no_replicate = TRUE)
```

Note that while dispersion estimation without replicates can be enabled by setting `no_replicate = TRUE`,
this approach is not recommended for biological experiments.
For reliable statistical inference,
we strongly encourage reconsidering the experimental design to include multiple biological replicates per group.


#### MCMC Options

Several options control the behavior of MCMC sampling and
can significantly affect the accuracy and efficiency of inference.
Although `hobit()` supports all options available in the `sample()` function
from the `r pkg('cmdstanr')` package,
the following are particularly important.

- `chains`: Number of MCMC chains to run (default: `4`).
- `iter_warmup`: Number of warmup iterations per chain (default: `1000`).
- `iter_sampling`: Number of post-warmup iterations per chain (default: `1000`).
- `save_warmup`: Whether to save warmup iterations (default: `FALSE`).
- `thin`: Interval at which samples are saved (default: `1`). Typically unchanged unless memory issues arise.
- `max_treedepth`: Maximum tree depth for the NUTS sampler (default: `10`). Increasing it may help resolve convergence issues but slows execution.
- `adapt_delta`: Target acceptance rate for the sampler (default: `0.8`). Higher values improve stability but may reduce speed.

The following example increases the number of chains and iterations to improve stability and convergence,
though at the cost of longer execution times:

```{r fc-hobit-mcmcopts-iters, eval=FALSE}
x_output <- hobit(x, chains = 8, iter_warmup = 2000, iter_sampling = 10000)
```

If warnings or convergence issues arise with the default settings,
consider increasing `max_treedepth` and `adapt_delta`.
Note that increasing these parameters may result in slower sampling.

```{r fc-hobit-mcmcopts-depth, eval=FALSE}
x_output <- hobit(x, max_treedepth = 12, adapt_delta = 0.95)
```

For more advanced configuration options, refer to the
[cmdstanr documentation](https://mc-stan.org/cmdstanr/reference/model-method-sample.html).


#### Verbosity Options

Verbosity options are available to control the verbosity of progress messages during sampling.

- `refresh`: Frequency of progress updates to the console (default: `100`). Set to `0` to suppress updates.
- `show_messages`: Whether to print all diagnostic and progress messages (default: `TRUE`).

To reduce console output during execution, adjust these settings as shown below.

```{r fc-hobit-vvopts, eval=FALSE}
x_output <- hobit(x, refresh = 0, show_messages = FALSE)
```



