## HomeoRoq

```{r fc-hq-setup, include=FALSE}
source("../_config.R")
```

HomeoRoq [@ref_homeoroq] estimates the probability that
homeolog expression ratios remain constant
between two experimental groups using RNA-seq read count data for each homeolog.
Since this probability cannot be derived analytically,
HomeoRoq adopts a Bayesian framework
to approximate the posterior distribution of homeolog expression ratios
under the null hypothesis of no change between conditions.
The `homeoroq()` function provides a complete interface
for detecting homeologs whose expression ratios differ between two conditions.

By default, it performs 10,000 sampling iterations (`iter_sampling = 1e4`) per probability calculation
and repeats the calculation four times (`chians = 4`).
However, for robust and reproducible estimation of statistical significance (i.e., *p*-values),
it is recommended to run multiple independent sampling replicates,
each with a sufficiently large number of iterations.
To reproduce the settings used in the original HomeoRoq implementation [@ref_homeoroq],
set `iter_sampling = 1e4` and `chains = 10`.


### Inputs

The `homeoroq()` function requires an input object of class `ExpMX`,
same to the `hobit()` function.


```{r fc-hq-input-newExpMX}
gexp <- read.table("../data/c_flexuosa.0516.mini.txt.gz",
                   header = TRUE, sep = "\t", row.names = 1)
group <- c("wet", "wet", "wet", "dry", "dry", "dry")
mapping_table <- read.table("../data/c_flexuosa.homeolog.txt.gz",
                            header = TRUE, sep = "\t")
x <- newExpMX(gexp, group, mapping_table)
```

### Test

Once an `ExpMX` object is prepared,
normalize the counts if the data have not already been normalized,
then run the test using the `homeoroq()` function.

```{r fc-hq-run-homeoroq, results="hide", message=FALSE, warning=FALSE}
x <- norm_counts(x)
x_output <- homeoroq(x)
```


### Outputs

The result (`x_output`) is a data frame with one row per homeolog,
summarizing the statistical test results.

```{r, fc-hq-output}
head(x_output)
```

- `gene`: Homeolog ID. The gene names defined in the first column of the mapping table.
- `pvalue`: *p*-value calculated from sampling.
- `qvalue`: Adjusted *p*-value using the Benjamini-Hochberg method.
- `sumexp__*__$`: Total read counts for subgenome `$` under condition `*`.
- `ratio__*__$`: Homeolog expression ratios for subgenome `$` under condition `*`.
- `ratio_sd`: Standard deviation of homeolog expression ratios calculated from observed counts.

The total read counts (`sumexp__*__$`) and homeolog expression ratios (`ratio__*__$`) 
are computed directly from the observed read counts.
Note that this differs from the `hobit()` function,
which calculates its statistics from MCMC samples.

Furthermore, the `pvalue` and `qvalue` columns in `x_output` may contain `NA` values.
This happens when at least one gene in a homeolog tuple is not expressed in one or more conditions,
making it impossible to calculate expression ratios.
As a result, the statistical test cannot be performed for those tuples.
These `NA` values may affect sorting or counting the number of significant homeolog tuples.
Therefore, we recommend replacing `NA` with `1` (i.e., non-significant)
before performing downstream analyses.
For example:


```{r fc-hq-output-print-na}
x_output$pvalue[is.na(x_output$pvalue)] <- 1
x_output$qvalue[is.na(x_output$qvalue)] <- 1
```



### Options

The number of iterations and number of samples drawn per iteration can be
adjusted using the `chains` and `iter_sampling` arguments, respectively.
For example:

```{r fc-hq-opts-iters, eval=FALSE}
x_output <- homeoroq(x, chains = 8, iter_sampling = 1e5)
```

Additionally, the `n_threads` option can be used to run iterations in parallel.

```{r fc-hq-opts-threds, eval=FALSE}
x_output <- homeoroq(x, chains = 8, iter_sampling = 1e5, n_threads = 8)
```

